{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0caf9543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log initialized. Log file: /tmp/das.log\n",
      "Atoms (nodes, links): (4381122, 9836656)\n"
     ]
    }
   ],
   "source": [
    "from das.distributed_atom_space import DistributedAtomSpace, QueryOutputFormat\n",
    "from das.pattern_matcher.pattern_matcher import PatternMatchingAnswer, OrderedAssignment, UnorderedAssignment, CompositeAssignment, Node, Link, Variable, Not, And, Or\n",
    "from das.database.db_interface import WILDCARD\n",
    "import warnings\n",
    "import time\n",
    "warnings.filterwarnings('ignore')\n",
    "das = DistributedAtomSpace()\n",
    "db = das.db\n",
    "print(\"Atoms (nodes, links): \" + str(das.count_atoms()))\n",
    "# 2023_02 :-->  ( 2584425,  26915329)\n",
    "# 2023_04 :-->  (11004962, 176057040)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ece77aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "class WallClock:\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.start_time = None\n",
    "        self.wall_time = None\n",
    "        self.epochs = 0\n",
    "    \n",
    "    def start(self):\n",
    "        self.start_time = time.perf_counter()\n",
    "    \n",
    "    def stop(self):\n",
    "        self.wall_time = time.perf_counter() - self.start_time\n",
    "    \n",
    "    def epoch(self, n=1):\n",
    "        self.epochs += n\n",
    "        \n",
    "    def print(self, text=\"query\"):\n",
    "        if self.wall_time >= 1:\n",
    "            total_time = f\"{self.wall_time:.3f} seconds\"\n",
    "        else:\n",
    "            total_time = f\"{(self.wall_time * 1000):.0f} milliseconds\"\n",
    "        if self.epochs == 0:\n",
    "            time_per_epoch = \"\"\n",
    "        else:\n",
    "            time_per_epoch = f\"{((self.wall_time * 1000) / self.epochs):.3f} milliseconds per {text}\"\n",
    "        print(f\"{total_time} ({time_per_epoch})\")\n",
    "            \n",
    "        \n",
    "def print_ordered_assignment(assignment):\n",
    "    if assignment is not None:\n",
    "        for key, value in assignment.mapping.items():\n",
    "            print(f\"{key}: {db.get_node_name(value)}\")\n",
    "\n",
    "def print_unordered_assignment(assignment):\n",
    "    if assignment is not None:\n",
    "        symbols = []\n",
    "        for key in assignment.symbols:\n",
    "            for i in range(assignment.symbols[key]):\n",
    "                symbols.append(key)\n",
    "        values = []\n",
    "        for key in assignment.values:\n",
    "            for i in range(assignment.values[key]):\n",
    "                values.append(key)\n",
    "        mapping_keys = []\n",
    "        mapping_values = []\n",
    "        for symbol, value in zip(symbols, values):\n",
    "            mapping_keys.append(symbol)\n",
    "            mapping_values.append(db.get_node_name(value))\n",
    "        print(f\"{mapping_keys} = {mapping_values}\")\n",
    "\n",
    "def print_elapsed_time(start):\n",
    "    end = time.perf_counter()\n",
    "    wall_time = end - start\n",
    "    if wall_time >= 1:\n",
    "        print(f\"{wall_time:.3f} seconds\")\n",
    "    else:\n",
    "        print(f\"{(wall_time * 1000):.0f} milliseconds\")\n",
    "        \n",
    "def query(query_obj, log = False, detailed_log = False):\n",
    "    assert log or (not detailed_log)\n",
    "    query_answer = PatternMatchingAnswer()\n",
    "    start = time.perf_counter()\n",
    "    matched = query_obj.matched(db, query_answer)\n",
    "    if log:\n",
    "        print_elapsed_time(start)\n",
    "        print(matched)\n",
    "        if matched:\n",
    "            print(f\"{len(query_answer.assignments)} answers\")\n",
    "            if detailed_log:\n",
    "        #         print(query_answer.assignments)\n",
    "                for assignment in query_answer.assignments:\n",
    "                    if type(assignment) is OrderedAssignment:\n",
    "                        print_ordered_assignment(assignment)\n",
    "                    elif type(assignment) is UnorderedAssignment:\n",
    "                        print_unordered_assignment(assignment)\n",
    "                    elif type(assignment) is CompositeAssignment:\n",
    "                        print_ordered_assignment(assignment.ordered_mapping)\n",
    "                        for unordered_assignment in assignment.unordered_mappings:\n",
    "                            print_unordered_assignment(unordered_assignment)\n",
    "                    print(\"\")\n",
    "    return query_answer.assignments\n",
    "\n",
    "def get_mappings(q, variable_name):\n",
    "    \"\"\"\n",
    "    Executes passed query and return the values assigned to the passed variable by searching for the respective node name\n",
    "    \"\"\"\n",
    "    assignments = query(q)\n",
    "    return [das.get_node_name(assignment.mapping[variable_name]) for assignment in assignments]\n",
    "    \n",
    "\n",
    "def get_feature_node_handle(name):\n",
    "    \"\"\"\n",
    "    Get the handle of the corresponding Gene node given a gene name.\n",
    "    \"\"\"\n",
    "    verbatim_node = das.get_node(\"Verbatim\", name)\n",
    "    schema_node = das.get_node(\"Schema\", \"Schema:feature_name\")\n",
    "    v1 = Variable(\"v1\")\n",
    "    links = das.get_links(\"Execution\", None, [schema_node, WILDCARD, verbatim_node])\n",
    "    # To be replaced by:\n",
    "    # assert len(links) > 0\n",
    "    if len(links) == 0:\n",
    "        return None\n",
    "    link = das.get_atom(links[0], output_format=QueryOutputFormat.ATOM_INFO)\n",
    "    return link[\"targets\"][1]\n",
    "\n",
    "\n",
    "def build_feature_node(name):\n",
    "    \"\"\"\n",
    "    Build a Node object to be used to compose queries. \n",
    "    \"\"\"\n",
    "    feature_node_handle = get_feature_node_handle(name)\n",
    "    if feature_node_handle == None:\n",
    "        return None\n",
    "    feature_node = das.get_atom(feature_node_handle, output_format=QueryOutputFormat.ATOM_INFO)\n",
    "    return Node(\"feature\", feature_node[\"name\"])\n",
    "\n",
    "\n",
    "def get_feature_fb_id(symbol):\n",
    "    \"\"\"\n",
    "    Get the FB id of a given feature by its symbol\n",
    "    \"\"\"\n",
    "    n = build_feature_node(symbol.replace(\"(\", \"[\").replace(\")\", \"]\"))\n",
    "    if n == None:\n",
    "        return None\n",
    "    v = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", \"Schema:feature_uniquename\")\n",
    "    # searches for an Execution link to the schema \"feature_uniquename\", node of type \"feature\" \n",
    "    # and value \"symbol\", and any ***atom***??? represented by Variable \"v1\"\n",
    "    q = Link(\"Execution\", ordered=True, targets=[s, n, v])\n",
    "    assignment = query(q)\n",
    "    assert len(assignment) == 1\n",
    "    id_handle = assignment.pop().mapping['v1'] # handle of \"Verbatim\" node\n",
    "    return db.get_node_name(id_handle)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    Get all handles gene node given a gene name (symbol).\n",
    "    \n",
    "    Some genes have more than  one uniquename. Eg \"AGO2\" in the next cell\n",
    "    Unfortunately, for some DAS setups, the uniquename retrieved by get_feature_fb_id doesn't work. \n",
    "    To ensure correct retrieval, the way is to get all FBgn# and test each one... :/\n",
    "\"\"\"    \n",
    "def get_all_feature_node_handles(name):\n",
    "    verbatim_node = das.get_node(\"Verbatim\", name)\n",
    "    schema_node = das.get_node(\"Schema\", \"Schema:feature_name\")\n",
    "    v1 = Variable(\"v1\")\n",
    "    links = das.get_links(\"Execution\", None, [schema_node, WILDCARD, verbatim_node])\n",
    "    # To be replaced by:\n",
    "    # assert len(links) > 0\n",
    "    if len(links) == 0:\n",
    "        return None\n",
    "    handles = []\n",
    "    for link in links:\n",
    "        atom_link = das.get_atom(link, output_format=QueryOutputFormat.ATOM_INFO)\n",
    "        handles.append(atom_link[\"targets\"][1])\n",
    "    return handles\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    Build a Node object to be used to compose queries. \n",
    "    \n",
    "    Some genes have more than  one uniquename. Eg \"AGO2\" in the next cell\n",
    "    Unfortunately, for some DAS setups, the uniquename retrieved by get_feature_fb_id doesn't work. \n",
    "    To ensure correct retrieval, the way is to get all FBgn# and test each one... :/\n",
    "\"\"\"\n",
    "def build_all_feature_nodes(name):\n",
    "    feature_node_handles = get_all_feature_node_handles(name)    \n",
    "    if feature_node_handles == None:\n",
    "        return None\n",
    "    nodes = []\n",
    "    for feature_node_handle in feature_node_handles:\n",
    "        feature_node = das.get_atom(feature_node_handle, output_format=QueryOutputFormat.ATOM_INFO)\n",
    "        nodes.append( Node(\"feature\", feature_node[\"name\"]) )\n",
    "    return nodes\n",
    "    \n",
    "\n",
    "\"\"\"\n",
    "    Some genes have more than  one uniquename. Eg \"AGO2\" in the next cell\n",
    "    Unfortunately, for some DAS setups, the uniquename retrieved by get_feature_fb_id doesn't work. \n",
    "    To ensure correct retrieval, the way is to get all FBgn# and test each one... :/\n",
    "\"\"\"\n",
    "def get_all_feature_fb_id(symbol):\n",
    "    nodes = build_all_feature_nodes(symbol.replace(\"(\", \"[\").replace(\")\", \"]\"))\n",
    "    if nodes == None:\n",
    "        return None\n",
    "    fb_ids = []\n",
    "    for node in nodes:\n",
    "        uniq_var = Variable(\"v1\")\n",
    "        schema = Node(\"Schema\", \"Schema:feature_uniquename\")\n",
    "        q = Link(\"Execution\", ordered=True, targets=[schema, node, uniq_var])\n",
    "        assignment = query(q)\n",
    "        assert len(assignment) == 1\n",
    "        id_handle = assignment.pop().mapping['v1'] # handle of \"Verbatim\" node\n",
    "        fb_ids.append( db.get_node_name(id_handle) )\n",
    "    return fb_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2e6bf581-af51-4329-8272-c237cc9ddab1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FBgn0003607  ::---->    Su(var)205\n",
      "FBgn0026015  ::---->    Top3beta\n",
      "FBgn0011656  ::---->    Mef2\n",
      "FBgn0023076  ::---->    Clk\n",
      "FBgn0015664  ::---->    Dref\n",
      "FBgn0004915  ::---->    TfIIB\n",
      "FBgn0262656  ::---->    Myc\n",
      "FBgn0087035  ::---->    AGO2\n",
      "FBgn0026401  ::---->    Nipped-B\n",
      "FBgn0000283  ::---->    Cp190\n",
      "FBgn0011289  ::---->    TfIIA-L\n",
      "FBgn0013263  ::---->    Trl\n",
      "FBgn0005386  ::---->    ash1\n",
      "FBgn0003079  ::---->    Raf\n",
      "FBgn0000015  ::---->    Abd-B\n",
      "FBgn0015270  ::---->    Orc2\n",
      "FBgn0015799  ::---->    Rbf\n",
      "FBgn0014340  ::---->    mof\n",
      "FBgn0005617  ::---->    msl-1\n",
      "FBgn0001206  ::---->    Hmr\n",
      "\n",
      "\n",
      "All FBgn# for gene Argonaut:\n",
      "['FBgn0087035', 'FBgn0046812']\n",
      "\n",
      "\n",
      "All FBgn# for gene cap-n-collar-RG:\n",
      "['FBtr0306750', 'FBtr0084394']\n",
      "\n",
      "\n",
      "All FBgn# for gene Diedel:\n",
      "['Q9VAK8-IPR025061-PF13164-FBpp0084794-1-tr', 'FBgn0039666', 'Q9VWM9-IPR025061-PF13164-FBpp0310706-1-tr', 'A0A0B4K6M8-IPR025061-PF13164-FBpp0297389-1-tr', 'Q9VWM9-IPR025061-PF13164-FBpp0310706-1-tr-exon0', 'A0A0B4K6M8-IPR025061-PF13164-FBpp0297389-1-tr-exon0', 'Q9VAK8-IPR025061-PF13164-FBpp0084794-1-tr-exon0']\n"
     ]
    }
   ],
   "source": [
    "# the FBgn# for the 20 NetAct TFs.\n",
    "# In this small \"GeneGroup DAS\" \"Nipped-B\" and \"Rbf\" are not in the \"feature_uniquename\" schema\n",
    "for name in [\"Su(var)205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \n",
    "             \"Cp190\", \"TfIIA-L\", \"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]:\n",
    "    fb_id = get_feature_fb_id(name)\n",
    "    print(str(fb_id) + \"  ::---->    \" + name)\n",
    "\n",
    "print(\"\\n\\nAll FBgn# for gene Argonaut:\")\n",
    "fb_ids = get_all_feature_fb_id(\"AGO2\")\n",
    "print (fb_ids)\n",
    "print(\"\\n\\nAll FBgn# for gene cap-n-collar-RG:\")\n",
    "fb_ids = get_all_feature_fb_id(\"cnc-RG\")\n",
    "print (fb_ids)\n",
    "print(\"\\n\\nAll FBgn# for gene Diedel:\")\n",
    "fb_ids = get_all_feature_fb_id(\"Diedel\")\n",
    "print (fb_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7549318-45bf-4c59-92e8-305b2c75921c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "58d5ecb3-b9bd-4c88-8701-0a8e577ccd90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# General building block\n",
    "\n",
    "def get_feature_entities_list(feature_identifier, another_entity_table_column_name, feature_identifier_type=\"symbol\"):\n",
    "    \"\"\"\n",
    "        Gets a ***list of entities*** from table_column \"another_entity_table_column_name\" [as stored in the \"Schema Node\" whose name is\n",
    "        exactly the same (\"another_entity_table_column_name\")] for a given feature (feature_identifier) \n",
    "        Most times the returned list will contain only one element.\n",
    "        Look the next cell for examples of using this query\n",
    "    \n",
    "        Args:\n",
    "            feature_identifer (str): string that identifies a feature. Most frequently it will be a gene_symbol or a uniquename (FBid).\n",
    "            another_entity_table_column_name (str): this parameter must be the SchemaNode's name (whithout the \"Schema:\" prefix) of the \n",
    "                                                    second element of the \"essential column pair\" that is related somehow to the \n",
    "                                                    feature_identifier parameter\n",
    "            feature_identifier_type (str, optional): string that indicates the type of the feature_identifier parameter. Most commonly it will be \n",
    "                                                     \"symbol\" (it is the default) that indicates a gene symbol\n",
    "                                                     \"FBid\" that indicates a current (valid, non obsolete) uniquename. Eg: FBgn0000166, FBpp0310706.\n",
    "                                                     \"self\" that indicates to use the value of the feature_identifier parameter\n",
    "            \n",
    "                \n",
    "        Returns:\n",
    "            A list of feature-related entities (strings). Most times it will hold only one element.\n",
    "    \"\"\"\n",
    "    if feature_identifier_type == \"symbol\":\n",
    "        return _get_feature_entities_list(feature_identifier, another_entity_table_column_name)\n",
    "    else:\n",
    "        if feature_identifier_type == \"FBid\" or feature_identifier_type == \"self\":\n",
    "            n1 = Node(\"Verbatim\", feature_identifier)\n",
    "            v1 = Variable(\"v1\")\n",
    "            s = Node(\"Schema\", f\"Schema:{another_entity_table_column_name.replace('(', '[').replace(')', ']')}\")    \n",
    "        \n",
    "            q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])    \n",
    "        else:\n",
    "            print(feature_identifier_type)\n",
    "    return get_mappings(q1, \"v1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d40ea22e-a863-4729-b4a1-44ea3e540582",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look the next cell for examples of using this query\n",
    "def _get_feature_entities_list(gene_symbol, another_entity_table_column_name):\n",
    "    fb_id = get_feature_fb_id(gene_symbol)     \n",
    "    if fb_id == None:\n",
    "        return None\n",
    "    n1 = Node(\"Verbatim\", fb_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", f\"Schema:{another_entity_table_column_name.replace('(', '[').replace(')', ']')}\")    \n",
    "\n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])    \n",
    "    return get_mappings(q1, \"v1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "070d3401-fc6a-4092-8ca0-4042fb4014bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0f58e7e1-4972-4539-8e4f-0bee96b5c532",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Gene Top3beta:\n",
      "Top3beta (fbgn_fbtr_fbpp_expanded) organism: ['Dmel']\n",
      "Top3beta (Dmel_enzyme_data) GO terms: ['GO:0003917']\n",
      "Top3beta (Dmel_enzyme_data) gene group id (FBgg#): ['FBgg0001554']\n",
      "Top3beta (Dmel_enzyme_data) gene_group_EC_number(s): ['EC 5.6.2.1']\n",
      "Top3beta (Dmel_enzyme_data) gene_EC_number(s): ['EC 5.6.2.1']\n",
      "Top3beta (Dmel_enzyme_data) gene_name: ['Topoisomerase 3beta']\n",
      "Top3beta (gene_map_table) recombination_loc: ['1-16']\n",
      "Top3beta (gene_map_table) cytogenetic_loc: ['5E5-5E5']\n",
      "Top3beta (dmel_human_orthologs_disease) Human_gene_symbol: []\n",
      "Top3beta (dmel_human_orthologs_disease) Dmel_gene_ID: ['FBgn0026015']\n",
      "Top3beta (dmel_human_orthologs_disease) Human_gene_symbol: ['TOP3B']\n",
      "Top3beta (dmel_human_orthologs_disease) Human_gene_symbol: ['TOP3A']\n",
      "Top3beta (dmel_human_orthologs_disease) Dmel_gene_symbol: ['Top3beta']\n",
      "Top3beta (best_gene_summary) Summary source: ['FlyBase Gene Snapshot']\n",
      "Top3beta (best_gene_summary) Summary: ['Topoisomerase 3beta [Top3beta] encodes an RNA topoisomerase that participates in both transcription and translation. It promotes transcription by being recruited by the product of Tdrd3, an epigenetic reader recognizing dimethylarginine motifs in core histones. It contributes to translation by complexing with the product of Fmr1, an RNA-binding protein deficient in fragile X mental retardation syndrome.']\n"
     ]
    }
   ],
   "source": [
    "    symbols_list = [\"Top3beta\", \"Mef2\", \"Clk\", \"Myc\", \"Abd-B\"]\n",
    "symbols_list = [\"Su(var)205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \n",
    "                \"Cp190\", \"TfIIA-L\",\"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]\n",
    "symbols_list = [\"Top3beta\"]               \n",
    "for gene_symbol in symbols_list:\n",
    "    print(f'\\n\\nGene {gene_symbol}:')\n",
    "    print(f'{gene_symbol} (fbgn_fbtr_fbpp_expanded) organism: {get_feature_entities_list(gene_symbol, \"fbgn_fbtr_fbpp_expanded_organism\")}')\n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) GO terms: {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_group_GO_id(s)\")}') \n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) gene group id (FBgg#): {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_group_id\")}') \n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) gene_group_EC_number(s): {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_group_EC_number(s)\")}') \n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) gene_EC_number(s): {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_EC_number(s)\")}')     \n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) gene_name: {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_name\")}')     \n",
    "    \n",
    "    print(f'{gene_symbol} (gene_map_table) recombination_loc: {get_feature_entities_list(gene_symbol, \"gene_map_table_recombination_loc\")}') \n",
    "    print(f'{gene_symbol} (gene_map_table) cytogenetic_loc: {get_feature_entities_list(gene_symbol, \"gene_map_table_cytogenetic_loc\")}') \n",
    "\n",
    "    print(f'{gene_symbol} (best_gene_summary) Summary source: {get_feature_entities_list(gene_symbol, \"best_gene_summary_Summary_Source\")}') \n",
    "    print(f'{gene_symbol} (best_gene_summary) Summary: {get_feature_entities_list(gene_symbol, \"best_gene_summary_Summary\")}') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ede269f-7f84-45c8-8e8b-8f973e2af7ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0b43bb69-2ef0-433b-9a5e-944f06872b5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Gene Raf:\n",
      "FBal0294413 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0294413 (allele_genetic_interactions) interaction: []\n",
      "FBal0013773 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0013773 (allele_genetic_interactions) interaction: []\n",
      "FBal0374656 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0374656 (allele_genetic_interactions) interaction: []\n",
      "FBal0245477 (allele_genetic_interactions) Allele symbol: ['Raf[JF01185]']\n",
      "FBal0245477 (allele_genetic_interactions) interaction: ['Raf[JF01185], Scer\\\\GAL4[tin.CΔ4] is a non-suppressor of adult heart phenotype of Scer\\\\GAL4[tin.CΔ4], yki[S168A.UAS.Tag:V5]']\n",
      "FBal0103602 (allele_genetic_interactions) Allele symbol: ['Raf[LE78]']\n",
      "FBal0103602 (allele_genetic_interactions) interaction: ['Raf[LE78] is a non-suppressor of phenotype of cic[1]']\n",
      "FBal0127909 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0127909 (allele_genetic_interactions) interaction: []\n",
      "FBal0372506 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0372506 (allele_genetic_interactions) interaction: []\n",
      "FBal0103603 (allele_genetic_interactions) Allele symbol: ['Raf[F20.UAS]']\n",
      "FBal0103603 (allele_genetic_interactions) interaction: ['Raf[F20.UAS], Scer\\\\GAL4[604] has wing vein L5 phenotype, suppressible by Sps1[k11320]/SelD[+]', 'Raf[F20.UAS], Scer\\\\GAL4[604] has wing phenotype, suppressible by Sps1[k11320]/SelD[+]', 'Raf[F20.UAS], Scer\\\\GAL4[604] has wing vein L3 phenotype, suppressible by Sps1[k11320]/SelD[+]']\n",
      "FBal0278625 (allele_genetic_interactions) Allele symbol: []\n",
      "FBal0278625 (allele_genetic_interactions) interaction: []\n",
      "FBal0282812 (allele_genetic_interactions) Allele symbol: ['Raf[DN.UAS.cUa]']\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Invalid handle: 5985381cefe9a1115d36d5f933b1a3e7",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mValueError\u001B[0m                                Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[78], line 17\u001B[0m\n\u001B[1;32m     13\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m allele_id \u001B[38;5;129;01min\u001B[39;00m alleles_ids: \n\u001B[1;32m     14\u001B[0m     \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mallele_id\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m (allele_genetic_interactions) Allele symbol: \u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[1;32m     15\u001B[0m           \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mget_feature_entities_list(allele_id,\u001B[38;5;250m \u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mallele_genetic_interactions_allele_symbol\u001B[39m\u001B[38;5;124m\"\u001B[39m,\u001B[38;5;250m \u001B[39mfeature_identifier_type\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mFBid\u001B[39m\u001B[38;5;124m\"\u001B[39m)\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m     16\u001B[0m     \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mallele_id\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m (allele_genetic_interactions) interaction: \u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[0;32m---> 17\u001B[0m          \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mget_feature_entities_list(allele_id,\u001B[38;5;250m \u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mallele_genetic_interactions_interaction\u001B[39m\u001B[38;5;124m\"\u001B[39m,\u001B[38;5;250m \u001B[39mfeature_identifier_type\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mFBid\u001B[39m\u001B[38;5;124m\"\u001B[39m)\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m     18\u001B[0m \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mgene_symbol\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m (dmel_human_orthologs_disease) OMIM_Phenotype_IDs[name]: \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mget_feature_entities_list(gene_symbol,\u001B[38;5;250m \u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mdmel_human_orthologs_disease_OMIM_Phenotype_IDs[name]\u001B[39m\u001B[38;5;124m\"\u001B[39m)\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m'\u001B[39m) \n\u001B[1;32m     19\u001B[0m \u001B[38;5;66;03m# THIS WOULD RETURN [] BECAUSE THERE IS NO PAIR FROM \"feature_uniquename\" TO \"dmel_human_orthologs_disease_Human_gene_symbol\"\u001B[39;00m\n\u001B[1;32m     20\u001B[0m \u001B[38;5;66;03m# BUT THERE IS FROM \"dmel_human_orthologs_disease_Dmel_gene_ID\" TO \"dmel_human_orthologs_disease_Human_gene_symbol\" THAT SHOULD BE USED\u001B[39;00m\n",
      "Cell \u001B[0;32mIn[28], line 35\u001B[0m, in \u001B[0;36mget_feature_entities_list\u001B[0;34m(feature_identifier, another_entity_table_column_name, feature_identifier_type)\u001B[0m\n\u001B[1;32m     33\u001B[0m     \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m     34\u001B[0m         \u001B[38;5;28mprint\u001B[39m(feature_identifier_type)\n\u001B[0;32m---> 35\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mget_mappings\u001B[49m\u001B[43m(\u001B[49m\u001B[43mq1\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mv1\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m)\u001B[49m\n",
      "Cell \u001B[0;32mIn[12], line 88\u001B[0m, in \u001B[0;36mget_mappings\u001B[0;34m(q, variable_name)\u001B[0m\n\u001B[1;32m     84\u001B[0m \u001B[38;5;250m\u001B[39m\u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[1;32m     85\u001B[0m \u001B[38;5;124;03mExecutes passed query and return the values assigned to the passed variable by searching for the respective node name\u001B[39;00m\n\u001B[1;32m     86\u001B[0m \u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[1;32m     87\u001B[0m assignments \u001B[38;5;241m=\u001B[39m query(q)\n\u001B[0;32m---> 88\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m [das\u001B[38;5;241m.\u001B[39mget_node_name(assignment\u001B[38;5;241m.\u001B[39mmapping[variable_name]) \u001B[38;5;28;01mfor\u001B[39;00m assignment \u001B[38;5;129;01min\u001B[39;00m assignments]\n",
      "Cell \u001B[0;32mIn[12], line 88\u001B[0m, in \u001B[0;36m<listcomp>\u001B[0;34m(.0)\u001B[0m\n\u001B[1;32m     84\u001B[0m \u001B[38;5;250m\u001B[39m\u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[1;32m     85\u001B[0m \u001B[38;5;124;03mExecutes passed query and return the values assigned to the passed variable by searching for the respective node name\u001B[39;00m\n\u001B[1;32m     86\u001B[0m \u001B[38;5;124;03m\"\"\"\u001B[39;00m\n\u001B[1;32m     87\u001B[0m assignments \u001B[38;5;241m=\u001B[39m query(q)\n\u001B[0;32m---> 88\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m [\u001B[43mdas\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_node_name\u001B[49m\u001B[43m(\u001B[49m\u001B[43massignment\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mmapping\u001B[49m\u001B[43m[\u001B[49m\u001B[43mvariable_name\u001B[49m\u001B[43m]\u001B[49m\u001B[43m)\u001B[49m \u001B[38;5;28;01mfor\u001B[39;00m assignment \u001B[38;5;129;01min\u001B[39;00m assignments]\n",
      "File \u001B[0;32m/app/das/distributed_atom_space.py:296\u001B[0m, in \u001B[0;36mDistributedAtomSpace.get_node_name\u001B[0;34m(self, node_handle)\u001B[0m\n\u001B[1;32m    295\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mget_node_name\u001B[39m(\u001B[38;5;28mself\u001B[39m, node_handle: \u001B[38;5;28mstr\u001B[39m) \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m>\u001B[39m \u001B[38;5;28mstr\u001B[39m:\n\u001B[0;32m--> 296\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdb\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_node_name\u001B[49m\u001B[43m(\u001B[49m\u001B[43mnode_handle\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/app/das/database/redis_mongo_db.py:284\u001B[0m, in \u001B[0;36mRedisMongoDB.get_node_name\u001B[0;34m(self, node_handle)\u001B[0m\n\u001B[1;32m    282\u001B[0m answer \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_retrieve_key_value(KeyPrefix\u001B[38;5;241m.\u001B[39mNAMED_ENTITIES, node_handle)\n\u001B[1;32m    283\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m answer:\n\u001B[0;32m--> 284\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mInvalid handle: \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mnode_handle\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m    285\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m answer[\u001B[38;5;241m0\u001B[39m]\u001B[38;5;241m.\u001B[39mdecode()\n",
      "\u001B[0;31mValueError\u001B[0m: Invalid handle: 5985381cefe9a1115d36d5f933b1a3e7"
     ]
    }
   ],
   "source": [
    "# Allele / diseases related gene data:\n",
    "symbols_list = [\"Top3beta\", \"Mef2\", \"Clk\", \"Myc\", \"Abd-B\"]\n",
    "symbols_list = [\"Su(var)205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \n",
    "                \"Cp190\", \"TfIIA-L\",\"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]\n",
    "symbols_list = [\"Top3beta\"]               \n",
    "symbols_list = [\"Raf\"]               \n",
    "#symbols_list = [\"Rbf\"]\n",
    "for gene_symbol in symbols_list:\n",
    "    print(f'\\n\\nGene {gene_symbol}:')\n",
    "    alleles_ids = get_feature_entities_list(gene_symbol, \"fbal_to_fbgn_AlleleID\")\n",
    "    \n",
    "    # relating another allele info...\t\t\t\tinteraction\t\tFBrf#     Top3β[hs.PP] is a non-suppressor of lethal | recessive phenotype of Top3α[191]\n",
    "    for allele_id in alleles_ids: \n",
    "        print(f'{allele_id} (allele_genetic_interactions) Allele symbol: '\n",
    "              f'{get_feature_entities_list(allele_id, \"allele_genetic_interactions_allele_symbol\", feature_identifier_type=\"FBid\")}')\n",
    "        print(f'{allele_id} (allele_genetic_interactions) interaction: '\n",
    "             f'{get_feature_entities_list(allele_id, \"allele_genetic_interactions_interaction\", feature_identifier_type=\"FBid\")}')\n",
    "        \n",
    "    print(f'{gene_symbol} (dmel_human_orthologs_disease) OMIM_Phenotype_IDs[name]: {get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_OMIM_Phenotype_IDs[name]\")}') \n",
    "    # THIS WOULD RETURN [] BECAUSE THERE IS NO PAIR FROM \"feature_uniquename\" TO \"dmel_human_orthologs_disease_Human_gene_symbol\"\n",
    "    # BUT THERE IS FROM \"dmel_human_orthologs_disease_Dmel_gene_ID\" TO \"dmel_human_orthologs_disease_Human_gene_symbol\" THAT SHOULD BE USED\n",
    "    print(f'{gene_symbol} (dmel_human_orthologs_disease) Human_gene_symbol: {get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_Human_gene_symbol\")}') \n",
    "    #print(f'{gene_symbol} (dmel_human_orthologs_disease) Dmel_gene_ID: {get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_Dmel_gene_ID\", feature_identifier_type=\"self\")}')     \n",
    "    human_HGNC_ids = get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_Human_gene_HGNC_ID\")\n",
    "    print(f'{gene_symbol} (dmel_human_orthologs_disease) Human_gene_HGNC_ID: {human_HGNC_ids}')\n",
    "    for hgnc_id in human_HGNC_ids:\n",
    "        print(f'{gene_symbol} (dmel_human_orthologs_disease) Human_gene_symbol: {get_feature_entities_list(hgnc_id, \"dmel_human_orthologs_disease_Human_gene_symbol\", feature_identifier_type=\"self\")}') \n",
    "    print(f'{gene_symbol} (dmel_human_orthologs_disease) Dmel_gene_symbol: {get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_Dmel_gene_symbol\")}')     \n",
    "    print(f'{gene_symbol} (dmel_human_orthologs_disease) Human_gene_OMIM_ID: {get_feature_entities_list(gene_symbol, \"dmel_human_orthologs_disease_Human_gene_OMIM_ID\")}') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81b633c-eeeb-4cdc-bcc9-ca24a06077d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "66758cdf-c5ac-48cf-8c6a-026eebeb8648",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evaluation \"Predicate has_name\" (List \"Enzyme EC 1.1.1.1\" \"Concept alcohol dehydrogenase\"))\n",
    "\n",
    "# (Execution \"Schema Schema:grp_uniquename\" \"grp 162\" \"Verbatim FBgg0000201\")\n",
    "# (Execution \"Schema Schema:public.cvterm\" \"grp 162\" \"Concept Concept:public.cvterm_505\")\n",
    "# (Execution \"Schema Schema:grp_uniquename\" \"grp 163\" \"Verbatim FBgg0000202\")\n",
    "# (Execution \"Schema Schema:public.cvterm\" \"grp 163\" \"Concept Concept:public.cvterm_505\")\n",
    "# (Execution \"Schema Schema:grp_is_analysis\" \"grp 163\" \"Concept Concept:False\")\n",
    "\n",
    "# (Evaluation \"Predicate has_name\" (List \"Enzyme EC 1.1.1.7\" \"Concept propanediol-phosphate dehydrogenase\"))\n",
    "#parent_query = Link(\"Evaluation\", ordered=True, targets=[group_node, parent_var])\n",
    "\n",
    "def get_EC_name(ec_number):\n",
    "    if ec_number.startswith(\"EC\"):\n",
    "        enz_node = Node(\"Enzyme\", ec_number)\n",
    "    else:\n",
    "        enz_node = Node(\"EnzymeOntology\", ec_number)\n",
    "    v1 = Variable(\"v1\")\n",
    "    has_name_pred_node = Node(\"Predicate\", \"has_name\")    \n",
    "\n",
    "    q1 = Link(\"Evaluation\", ordered=True, targets=[has_name_pred_node, enz_node, v1])    \n",
    "    return get_mappings(q1, \"v1\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "581c4747-062e-47f3-90aa-f48d354b653e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "8f8afe44-13c0-4765-8371-554ed76da248",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'dict_data/go-namespace.json'",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mFileNotFoundError\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[76], line 4\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;28;01mimport\u001B[39;00m \u001B[38;5;21;01mjson\u001B[39;00m\n\u001B[1;32m      3\u001B[0m \u001B[38;5;66;03m# GO dictionary for knowing correct GO nodes types\u001B[39;00m\n\u001B[0;32m----> 4\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m \u001B[38;5;28;43mopen\u001B[39;49m\u001B[43m(\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mdict_data/go-namespace.json\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mr\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m)\u001B[49m \u001B[38;5;28;01mas\u001B[39;00m ns:\n\u001B[1;32m      5\u001B[0m     go_plus_dict \u001B[38;5;241m=\u001B[39m json\u001B[38;5;241m.\u001B[39mload(ns)\n\u001B[1;32m      8\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mget_GO_term_name\u001B[39m(go_term):\n\u001B[1;32m      9\u001B[0m     \u001B[38;5;66;03m# finds go_term type (BP, MF, CC)\u001B[39;00m\n",
      "File \u001B[0;32m/usr/local/lib/python3.9/site-packages/IPython/core/interactiveshell.py:308\u001B[0m, in \u001B[0;36m_modified_open\u001B[0;34m(file, *args, **kwargs)\u001B[0m\n\u001B[1;32m    301\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m file \u001B[38;5;129;01min\u001B[39;00m {\u001B[38;5;241m0\u001B[39m, \u001B[38;5;241m1\u001B[39m, \u001B[38;5;241m2\u001B[39m}:\n\u001B[1;32m    302\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\n\u001B[1;32m    303\u001B[0m         \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mIPython won\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mt let you open fd=\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mfile\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m by default \u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    304\u001B[0m         \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mas it is likely to crash IPython. If you know what you are doing, \u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    305\u001B[0m         \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124myou can use builtins\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124m open.\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    306\u001B[0m     )\n\u001B[0;32m--> 308\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mio_open\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfile\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n",
      "\u001B[0;31mFileNotFoundError\u001B[0m: [Errno 2] No such file or directory: 'dict_data/go-namespace.json'"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "# GO dictionary for knowing correct GO nodes types\n",
    "with open(\"dict_data/go-namespace.json\", \"r\") as ns:\n",
    "    go_plus_dict = json.load(ns)\n",
    "\n",
    "   \n",
    "def get_GO_term_name(go_term):\n",
    "    # finds go_term type (BP, MF, CC)\n",
    "    if go_term.startswith('GO'):        # Gene Ontology ID\n",
    "        go_namespace = go_plus_dict[go_term]\n",
    "        #print(go_namespace)\n",
    "    go_node = Node(go_namespace, go_term)\n",
    "    v1 = Variable(\"v1\")\n",
    "    has_name_pred_node = Node(\"Predicate\", \"has_name\")    \n",
    "\n",
    "    q1 = Link(\"Evaluation\", ordered=True, targets=[has_name_pred_node, go_node, v1])    \n",
    "    return get_mappings(q1, \"v1\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ede7aea-1522-4e1f-b961-8b603b46e3fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "165c5156-d201-4ff5-9174-87e20dda6d0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Gene Top3beta:\n",
      "Top3beta (Dmel_enzyme_data) gene_group_EC_number(s): ['EC 5.6.2.1']\n",
      "Top3beta (Dmel_enzyme_data) gene_EC_number(s)/(explorEnz has_name): EC 5.6.2.1 / []\n",
      "Top3beta (Dmel_enzyme_data) GO term: GO:0003917\n"
     ]
    }
   ],
   "source": [
    "#  GO terms / predicates\n",
    "# ExplorEnz number / names\n",
    "\n",
    "symbols_list = [\"Top3beta\", \"Mef2\", \"Clk\", \"Myc\", \"Abd-B\"]\n",
    "symbols_list = [\"Su(var)205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \n",
    "                \"Cp190\", \"TfIIA-L\",\"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]\n",
    "symbols_list = [\"Top3beta\"]               \n",
    "for gene_symbol in symbols_list:\n",
    "    print(f'\\n\\nGene {gene_symbol}:') \n",
    " \n",
    "    print(f'{gene_symbol} (Dmel_enzyme_data) gene_group_EC_number(s): {get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_group_EC_number(s)\")}')\n",
    "    ec_numbers = get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_EC_number(s)\")\n",
    "    for ec_number in ec_numbers:\n",
    "        print(f'{gene_symbol} (Dmel_enzyme_data) gene_EC_number(s)/(explorEnz has_name): {ec_number} / {get_EC_name(ec_number)}')\n",
    "        \n",
    "    go_terms = get_feature_entities_list(gene_symbol, \"Dmel_enzyme_data_gene_group_GO_id(s)\")\n",
    "    for go_term in go_terms:\n",
    "        #print(f'{gene_symbol} (Dmel_enzyme_data) GO term: {get_GO_term_name(go_term)}')    \n",
    "        print(f'{gene_symbol} (Dmel_enzyme_data) GO term: {go_term}')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f7bbd51-8224-493c-bb1c-5d2fdcc32427",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b9bd77-8dc9-4450-aead-edcee2c96d92",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "88b8f6f3-0e69-4afe-b493-fdab64a28515",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "About HGNC groups. \n",
    "The \"HGNC_family_ID\" (eg 705) is a group id; not a gene id (which is like HGNC:705)\n",
    "Flybase assigned 705 to the PTP (PROTEIN TYROSINE PHOSPHATASES) group that corresponds to this: \n",
    "https://www.genenames.org/data/genegroup/#!/group/705 (Class I classical Cys-based phosphatases)\n",
    "\n",
    "So,the get_groups_HGNC_ids method returns a list of pairs (\"id\", id_URL) because Flybase states that could exist\n",
    "more than one HGNC id for a given Flybase group and the table \"gene_groups_HGNC_fb*.tsv\" reflects that statement.\n",
    "\n",
    "get_groups_HGNC_ids method output:\n",
    "\n",
    "For the current example (Flybase PTP group) it is:\n",
    "\n",
    "[(705, https://www.genenames.org/data/genegroup/#!/group/705)]\n",
    "\n",
    "\n",
    "For multiple HGNC ids with input of \"FBgg0000112\":\n",
    "\n",
    "FB_group_id\tFB_group_symbol\tFB_group_name\tHGNC_family_ID\n",
    "FBgg0000112\tINX\t            INNEXINS\t    314\n",
    "FBgg0000112\tINX\t            INNEXINS\t    288\n",
    "\n",
    "it outputs:\n",
    "\n",
    "[ \n",
    "    (314, https://www.genenames.org/data/genegroup/#!/group/314), \n",
    "    (288, https://www.genenames.org/data/genegroup/#!/group/288)\n",
    "]\n",
    "\"\"\"\n",
    "# 1\n",
    "# Gets the group ***symbol*** from the \"group_id\" (FBgg#) as stored in the \"Schema Node\" whose name is \"table_name\".\n",
    "def get_group_symbol(group_id, table_name):\n",
    "    n1 = Node(\"Verbatim\", group_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_symbol\")\n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "    return get_mappings(q1, \"v1\")[0]\n",
    "\n",
    "# 2\n",
    "# Gets the group ***name*** from the \"group_id\" (FBgg#) as stored in the \"Schema Node\" whose name is \"table_name\".\n",
    "def get_group_name(group_id, table_name):\n",
    "    n1 = Node(\"Verbatim\", group_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_name\")\n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "    return get_mappings(q1, \"v1\")[0]\n",
    "\n",
    "# 3\n",
    "# Gets a ***list*** of HGNC ids for the group designated by \"group_id\".\n",
    "def get_groups_HGNC_ids(group_id):\n",
    "    n1 = Node(\"Verbatim\", group_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", f\"Schema:gene_groups_HGNC_HGNC_family_ID\")\n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "    ids = get_mappings(q1, \"v1\")\n",
    "    HGNC_ids_data_list = []\n",
    "    for id in ids:\n",
    "        HGNC_ids_data_list.append((id, f\"https://www.genenames.org/data/genegroup/#!/group/{id}\"))\n",
    "    return HGNC_ids_data_list\n",
    "\n",
    "# 4\n",
    "# Gets a ***list of groups ids*** (FBgg#) from the \"group_symbol\" as stored in the \"Schema Node\" whose name is \"table_name\".    \n",
    "def get_groups_ids(gene_symbol, table_name):\n",
    "    fb_id = get_feature_fb_id(gene_symbol)     \n",
    "    if fb_id == None:\n",
    "        return None\n",
    "    n1 = Node(\"Verbatim\", fb_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_id\")    \n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])    \n",
    "    return get_mappings(q1, \"v1\")\n",
    "\n",
    "# 5\n",
    "# gets group id(s) (FBgg#) of the gene designated by \"gene_symbol\"\n",
    "def get_gene_groups_ids(gene_symbol):\n",
    "    return get_groups_ids(gene_symbol, \"gene_group_data\")\n",
    "\n",
    "# 6\n",
    "# gets PATHWAY group id(s) (FBgg#) of the gene designated by \"gene_symbol\"\n",
    "def get_pathway_groups_ids(gene_symbol):\n",
    "    return get_groups_ids(gene_symbol, \"pathway_group_data\")    \n",
    "    \n",
    "\n",
    "# 7\n",
    "# Gets a ***list of groups symbols*** from the \"generoup_symbol\" as stored in the \"Schema Node\" whose name is \"table_name\".    \n",
    "def get_groups_symbols(gene_symbol, table_name):\n",
    "    gene_groups_ids = get_groups_ids(gene_symbol, table_name)\n",
    "    if gene_groups_ids == None:\n",
    "        return None\n",
    "    gg_symbols = []\n",
    "    for gg_id in gene_groups_ids:\n",
    "        n1 = Node(\"Verbatim\", gg_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        # groups\n",
    "        s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_symbol\")\n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        for symb in get_mappings(q1, \"v1\"):\n",
    "            gg_symbols.append(symb)    \n",
    "    return gg_symbols\n",
    "\n",
    "# 8\n",
    "def get_gene_groups_symbols(gene_symbol):\n",
    "    return get_groups_symbols(gene_symbol, \"gene_group_data\")\n",
    "\n",
    "# 9\n",
    "def get_pathway_groups_symbols(gene_symbol):\n",
    "    return get_groups_symbols(gene_symbol, \"pathway_group_data\")\n",
    "\n",
    "\n",
    "# 10\n",
    "# get group name(s) of gene designated by gene_symbol\n",
    "def get_groups_names(gene_symbol, table_name):\n",
    "    gene_groups_ids = get_groups_ids(gene_symbol, table_name)   \n",
    "    if gene_groups_ids == None:\n",
    "        return None\n",
    "    gg_names = []\n",
    "    for gg_id in gene_groups_ids:\n",
    "        n1 = Node(\"Verbatim\", gg_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        # groups\n",
    "        s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_name\")        \n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        for gg_name in get_mappings(q1, \"v1\"):\n",
    "            gg_names.append(gg_name)\n",
    "    return gg_names\n",
    "\n",
    "# 11\n",
    "def get_gene_groups_names(gene_symbol):\n",
    "    return get_groups_names(gene_symbol, \"gene_group_data\")\n",
    "\n",
    "# 12\n",
    "def get_pathway_groups_names(gene_symbol):\n",
    "    return get_groups_names(gene_symbol, \"pathway_group_data\")\n",
    "\n",
    "# 13\n",
    "# In fact, a group could have more than one parent! E.g: FBgg0000275\n",
    "def get_parent_groups_ids(group_id):\n",
    "    group_node = Node(\"Verbatim\", group_id)\n",
    "    parent_var = Variable(\"v1\")\n",
    "    parent_query = Link(\"Inheritance\", ordered=True, targets=[group_node, parent_var])\n",
    "    return get_mappings(parent_query, \"v1\")\n",
    "\n",
    "\n",
    "# 14\n",
    "def get_parent_groups_symbols(group_id, table_name):\n",
    "    parent_ids = get_parent_groups_ids(group_id)\n",
    "    parents_symbols = []\n",
    "    for parent_id in parent_ids:\n",
    "        n1 = Node(\"Verbatim\", parent_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        s = Node(\"Schema\", f\"Schema:{table_name}_Parent_FB_group_symbol\")        \n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        parents_symbols.extend(get_mappings(q1, \"v1\"))    \n",
    "    return parents_symbols\n",
    "\n",
    "# 15    \n",
    "# In fact, a group could have more than one parent! E.g: FBgg0000275\n",
    "def get_parent_gene_groups_symbols(group_id):\n",
    "    return get_parent_groups_symbols(group_id, \"gene_group_data\")        \n",
    "\n",
    "# 16\n",
    "# In fact, a group could have more than one parent! E.g: FBgg0000275\n",
    "def get_parent_pathway_group_symbols(group_id):\n",
    "    return get_parent_groups_symbols(group_id, \"pathway_group_data\")  \n",
    "\n",
    "\n",
    "\n",
    "# 17\n",
    "# In fact, a group could have more than one parent. E.g: FBgg0000275\n",
    "# In the precomputed tables there are only group PARENT id and symbol...\n",
    "# So, the parent group names should be retrieved as the regular groups names.\n",
    "def get_parent_groups_names(group_id, table_name):\n",
    "    parent_ids = get_parent_groups_ids(group_id)\n",
    "    parents_names = []\n",
    "    for parent_id in parent_ids:\n",
    "        n1 = Node(\"Verbatim\", parent_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        s = Node(\"Schema\", f\"Schema:{table_name}_FB_group_name\")        \n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        parents_names.extend(get_mappings(q1, \"v1\"))    \n",
    "    return parents_names\n",
    "\n",
    "# 18\n",
    "# In fact, a group could have more than one parent! E.g: FBgg0000275\n",
    "def get_parent_gene_groups_names(group_id):\n",
    "    return get_parent_groups_names(group_id, \"gene_group_data\")\n",
    "\n",
    "# 19\n",
    "# In fact, a group could have more than one parent! E.g: FBgg0000275\n",
    "def get_parent_pathway_groups_names(group_id):\n",
    "    return get_parent_groups_names(group_id, \"pathway_group_data\")\n",
    "\n",
    "\n",
    "\n",
    "# 20\n",
    "def get_groups_members(gene_symbol, table_name):\n",
    "    gene_groups_ids = get_groups_ids(gene_symbol, table_name)\n",
    "    if gene_groups_ids == None:\n",
    "        return None\n",
    "    gene_group_symbols = []\n",
    "    fb_id_schema_name = f\"Schema:{table_name}_Group_member_FB_gene_id\"\n",
    "    gene_sym_chema_name = f\"Schema:{table_name}_Group_member_FB_gene_symbol\"\n",
    "    for gg_id in gene_groups_ids:\n",
    "        a_gg_symbols = []\n",
    "        n1 = Node(\"Verbatim\", gg_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        # gets all gene ids (FBgn#) of group given by gg_id\n",
    "        sp = Node(\"Schema\", fb_id_schema_name)\n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[sp, n1, v1])\n",
    "        #gene_ids = sorted(get_mappings(q1, \"v1\"))\n",
    "        gene_ids = sorted(list(set(get_mappings(q1, \"v1\"))))\n",
    "        for gene_id in gene_ids:\n",
    "            n2 = Node(\"Verbatim\", gene_id)\n",
    "            v2 = Variable(\"v2\")\n",
    "            ss = Node(\"Schema\", gene_sym_chema_name)\n",
    "            q1 = Link(\"Execution\", ordered=True, targets=[ss, n2, v2])\n",
    "            a_gg_symbols.append(get_mappings(q1, \"v2\")[0].replace(\"[\", \"(\").replace(\"]\", \")\"))\n",
    "        gene_group_symbols.append(((gg_id, get_group_symbol(gg_id, table_name), get_group_name(gg_id, table_name), len(a_gg_symbols), sorted(a_gg_symbols))))        \n",
    "    return gene_group_symbols\n",
    "\n",
    "# 21    \n",
    "# Returns list(s) of gene symbols that are members of the same group(s) of \n",
    "# the gene desinated by gene_symbol\n",
    "def get_gene_groups_members(gene_symbol):\n",
    "    return get_groups_members(gene_symbol, \"gene_group_data\")\n",
    "\n",
    "# 22\n",
    "# same as get_gene_group_members() for pathway groups\n",
    "def get_pathway_groups_members(gene_symbol):\n",
    "    return get_groups_members(gene_symbol, \"pathway_group_data\")\n",
    "\n",
    "\n",
    "\n",
    "# 23\n",
    "\"\"\"\n",
    "    Level zero holds the groups ids that the gene belongs to.\n",
    "    Upper levels hold the hierarchy itself.\n",
    "    An empty list signals that the hierarchy finished in the previous level (exception to zero\n",
    "    that means there is no hierarchy at all).\n",
    "\"\"\"\n",
    "def get_group_hierarchy(gene_symbol, table_name):\n",
    "    gene_group_ids = get_groups_ids(gene_symbol, table_name)\n",
    "    if gene_group_ids == None:\n",
    "        return None\n",
    "    level = 0\n",
    "    group_hierarchy = [(level, gene_group_ids)]    \n",
    "    while gene_group_ids != []:\n",
    "        end = []\n",
    "        for gg_id in gene_group_ids:\n",
    "            parents = get_parent_groups_ids(gg_id)\n",
    "            if parents != []:\n",
    "                group_hierarchy.append((level + 1, parents))\n",
    "                end.extend(parents)\n",
    "            else:\n",
    "                group_hierarchy.append((level + 1, parents))  # empty parents marks the end of the hierarchy\n",
    "        level += 1\n",
    "        gene_group_ids = end\n",
    "    return group_hierarchy\n",
    "\n",
    "# 24\n",
    "\"\"\"\n",
    "    Level zero holds the groups ids that the gene belongs to.\n",
    "    Upper levels hold the hierarchy itself.\n",
    "    An empty list signals that the hierarchy finished in the previous level (exception to zero\n",
    "    that means there is no hierarchy at all).\n",
    "\"\"\"\n",
    "def get_gene_group_hierarchy(gene_symbol):\n",
    "    return get_group_hierarchy(gene_symbol, \"gene_group_data\")\n",
    "\n",
    "# 25\n",
    "\"\"\"\n",
    "    Level zero holds the groups ids that the gene belongs to.\n",
    "    Upper levels hold the hierarchy itself.\n",
    "    An empty list signals that the hierarchy finished in the previous level (exception to zero\n",
    "    that means there is no hierarchy at all).\n",
    "\"\"\"\n",
    "def get_pathway_group_hierarchy(gene_symbol):\n",
    "    return get_group_hierarchy(gene_symbol, \"pathway_group_data\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "abd3455c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.781 seconds\n",
      "True\n",
      "736654 answers\n"
     ]
    }
   ],
   "source": [
    "# Get uniquename of all genes in the knowledge base\n",
    "v1 = Variable(\"v1\")\n",
    "v2 = Variable(\"v2\")\n",
    "#s = Node(\"Schema\", \"Schema:feature_uniquename\")\n",
    "s = Node(\"Schema\", \"Schema:feature_uniquename\")\n",
    "q1 = Link(\"Execution\", ordered=True, targets=[s, v1, v2])\n",
    "assignments = query(q1, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ca8b918a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FBgn0000015 -------------> Abd-B\n",
      "FBgn0023076 -------------> Clk\n",
      "FBgn0015799 -------------> Rbf\n",
      "FBgn0026015 -------------> Top3beta\n",
      "FBgn0005386 -------------> ash1\n",
      "FBgn0013263 -------------> Trl\n",
      "FBgn0026401 -------------> Nipped-B\n",
      "FBgn0015664 -------------> Dref\n",
      "FBgn0015270 -------------> Orc2\n",
      "FBgn0001206 -------------> Hmr\n",
      "FBgn0262656 -------------> Myc\n",
      "FBgn0003079 -------------> Raf\n",
      "FBgn0000283 -------------> Cp190\n",
      "FBgn0046812 -------------> AGO2\n",
      "FBgn0011289 -------------> TfIIA-L\n",
      "FBgn0087035 -------------> AGO2\n",
      "FBgn0004915 -------------> TfIIB\n",
      "FBgn0005617 -------------> msl-1\n",
      "FBgn0014340 -------------> mof\n",
      "FBgn0011656 -------------> Mef2\n",
      "1010.240 seconds (1.371 milliseconds per query)\n"
     ]
    }
   ],
   "source": [
    "# Print the mapping uniquename -> FB id for all features from the above query\n",
    "clock = WallClock()\n",
    "clock.start()\n",
    "for assignment in assignments:\n",
    "    #if clock.epochs > 100:\n",
    "        #break    \n",
    "    pkey_handle = assignment.mapping[\"v1\"] # handle of a \"gene\" node\n",
    "    unique_name_handle = assignment.mapping[\"v2\"] # handle of a \"Verbatim\" node\n",
    "    pkey = db.get_node_name(pkey_handle) # sequential integer used as PK in the DB table\n",
    "    unique_name = db.get_node_name(unique_name_handle) # FB id of the gene\n",
    "    v1 = Variable(\"v1\")\n",
    "    s = Node(\"Schema\", \"Schema:feature_name\")\n",
    "    q = Link(\"Execution\", ordered=True, targets=[s, Node(\"feature\", pkey), v1])\n",
    "    assignment2 = query(q)\n",
    "    assert len(assignment2) == 1 # There's only one link between the gene and its name\n",
    "    name_handle = assignment2.pop().mapping['v1'] # handle of \"Verbatim\" node\n",
    "    name = db.get_node_name(name_handle) # gene's name\n",
    "    if name in [\"Su[var]205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \"Cp190\", \"TfIIA-L\",\n",
    "               \"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]:\n",
    "        print(f\"{unique_name} -------------> {name}\")\n",
    "\n",
    "    clock.epoch()\n",
    "clock.stop()\n",
    "clock.print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6c324560-684f-4908-9c5d-d12f364c3a0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#from hyperon_das_atomdb import UNORDERED_LINK_TYPES, WILDCARD\n",
    "\n",
    "#from hyperon_das.api import DistributedAtomSpace\n",
    "#from hyperon_das.exceptions import QueryParametersException\n",
    "#from hyperon_das.pattern_matcher.pattern_matcher import And, Link, Variable\n",
    "#import hyperon_das\n",
    "import csv\n",
    "\n",
    "class DASLoadingVerifier:\n",
    "\n",
    "    '''\n",
    "        Generates a list like this:\n",
    "        table_strings = [\n",
    "            \"organism_abbreviation\",\n",
    "            \"synonym_name\",\n",
    "            \"gene_group_data_FB_group_id\",\n",
    "            \"gene_group_data_FB_group_symbol\" ,\n",
    "            \"gene_group_data_FB_group_name\"\n",
    "        ]\n",
    "\n",
    "        This is the name of \"Schema nodes\" in DAS lingo. A \"Schema node\" represents a column in\n",
    "        a Flybase precomputed table and it links all data in that column. The name of a Schema\n",
    "        node is composed of the \"underscored\" table name underscore \"underscored column name\". \n",
    "        An \"underscored name\" is a name in which the blank spaces were replced by underscores.\n",
    "\n",
    "        Parameter:  file_absolute_name\n",
    "                    Look the \"precomputed\" directory for the \"essential_pairs.txt\" file. It should\n",
    "                    be passed to this function.\n",
    "    '''\n",
    "    def read_precomputed_table_columns(self, file_absolute_name):\n",
    "        #with open(PRECOMPUTED_DIR + \"/essential_pairs.txt\", 'r') as file:\n",
    "        with open(file_absolute_name, 'r') as file:\n",
    "            reader = csv.reader(file, delimiter='\\t')\n",
    "            schema_nodes_list = []\n",
    "            for row in reader:\n",
    "                # print(row)\n",
    "                table_name = row[0].strip()\n",
    "                table_name = table_name.replace(' ', '_')\n",
    "                column_names = row[1:]      #  repetitions can be stored in this list, so\n",
    "                column_names = list( set( column_names ) )  # remove duplicates\n",
    "\n",
    "                for i in range(0, len(column_names)):\n",
    "                    if column_names[i] != '':\n",
    "                        schema_nodes_list.append( table_name + \"_\" + column_names[i].strip().replace(' ', '_') )\n",
    "\n",
    "        return schema_nodes_list\n",
    "\n",
    "\n",
    "    def print_schema_nodes_data(self, file_absolute_name, print_full_data=False):\n",
    "        #schema_names = self.read_precomputed_table_columns(file_absolute_name)\n",
    "        schema_names = [\n",
    "        \"ncRNA_genes_soTermId\", \"ncRNA_genes_primaryId\", \"ncRNA_genes_gene_symbol\", \"ncRNA_genes_gene_geneId\", \"ncRNA_genes_symbol\",\n",
    "        \"ncRNA_genes_gene_locusTag\", \"ncRNA_genes_taxonId\", \"ncRNA_genes_synonyms_symbol2\", \"ncRNA_genes_synonyms_symbol1\",\n",
    "        \"ncRNA_genes_cross_references_symbol2\", \"ncRNA_genes_cross_references_symbol1\", \"ncRNA_genes_related_sequences_sequenceId\",\n",
    "        \"ncRNA_genes_related_sequences_relationship\", \"ncRNA_genes_related_sequences_primaryId\", \"ncRNA_genes_gene_synonyms_symbol2\",\n",
    "        \"ncRNA_genes_gene_synonyms_symbol1\", \"ncRNA_genes_publications_publication\", \"ncRNA_genes_publications_primaryId\",\n",
    "        \"ncRNA_genes_genome_locations_primaryId\", \"ncRNA_genes_genome_locations_endPosition\", \"ncRNA_genes_genome_locations_assembly\",\n",
    "        \"ncRNA_genes_genome_locations_gca_accession\", \"ncRNA_genes_genome_locations_INSDC_accession\", \"ncRNA_genes_genome_locations_startPosition\",\n",
    "        \"ncRNA_genes_genome_locations_strand\", \"ncRNA_genes_genome_locations_chromosome\", \"fbgn_fbtr_fbpp_expanded_gene_symbol\",\n",
    "        \"fbgn_fbtr_fbpp_expanded_polypeptide_symbol\", \"fbgn_fbtr_fbpp_expanded_transcript_ID\", \"fbgn_fbtr_fbpp_expanded_transcript_symbol\",\n",
    "        \"fbgn_fbtr_fbpp_expanded_transcript_type\", \"fbgn_fbtr_fbpp_expanded_annotation_ID\", \"fbgn_fbtr_fbpp_expanded_organism\",\n",
    "        \"fbgn_fbtr_fbpp_expanded_polypeptide_ID\", \"fbgn_fbtr_fbpp_expanded_gene_fullname\", \"fbgn_fbtr_fbpp_expanded_gene_ID\",\n",
    "        \"dmel_human_orthologs_disease_Dmel_gene_symbol\", \"dmel_human_orthologs_disease_Human_gene_HGNC_ID\",\n",
    "        \"dmel_human_orthologs_disease_Human_gene_OMIM_ID\", \"dmel_human_orthologs_disease_Dmel_gene_ID\",\n",
    "        \"dmel_human_orthologs_disease_Human_gene_symbol\", \"dmel_human_orthologs_disease_DIOPT_score\",\n",
    "        \"dmel_human_orthologs_disease_OMIM_Phenotype_IDs[name]\", \"Dmel_enzyme_data_gene_symbol\", \"Dmel_enzyme_data_gene_EC_number(s)\",\n",
    "        \"Dmel_enzyme_data_gene_group_EC_number(s)\", \"Dmel_enzyme_data_gene_name\", \"Dmel_enzyme_data_gene_group_name\",\n",
    "        \"Dmel_enzyme_data_gene_group_GO_id(s)\", \"Dmel_enzyme_data_gene_id\", \"Dmel_enzyme_data_gene_group_id\",\n",
    "        \"gene_groups_HGNC_FB_group_name\", \"gene_groups_HGNC_FB_group_symbol\", \"gene_groups_HGNC_FB_group_id\",\n",
    "        \"gene_groups_HGNC_HGNC_family_ID\", \"best_gene_summary_Summary\", \"best_gene_summary_Summary_Source\",\n",
    "        \"best_gene_summary_Gene_Symbol\", \"best_gene_summary_FBgn_ID\", \"gene_map_table_cytogenetic_loc\",\n",
    "        \"gene_map_table_organism_abbreviation\", \"gene_map_table_recombination_loc\", \"gene_map_table_primary_FBid\",\n",
    "        \"gene_map_table_current_symbol\", \"gene_map_table_sequence_loc\", \"pathway_group_data_Parent_FB_group_symbol\",\n",
    "        \"pathway_group_data_Parent_FB_group_id\", \"pathway_group_data_Group_member_FB_gene_id\", \"pathway_group_data_FB_group_name\",\n",
    "        \"pathway_group_data_Group_member_FB_gene_symbol\", \"pathway_group_data_FB_group_symbol\", \"pathway_group_data_FB_group_id\",\n",
    "        \"genotype_phenotype_data_qualifier_ids\", \"genotype_phenotype_data_genotype_symbols\", \"genotype_phenotype_data_qualifier_names\",\n",
    "        \"genotype_phenotype_data_phenotype_id\", \"genotype_phenotype_data_genotype_FBids\", \"genotype_phenotype_data_phenotype_name\",\n",
    "        \"genotype_phenotype_data_reference\", \"gene_group_data_Parent_FB_group_symbol\", \"gene_group_data_Parent_FB_group_id\",\n",
    "        \"gene_group_data_Group_member_FB_gene_id\", \"gene_group_data_FB_group_name\", \"gene_group_data_Group_member_FB_gene_symbol\",\n",
    "        \"gene_group_data_FB_group_symbol\", \"gene_group_data_FB_group_id\", \"dmel_unique_protein_isoforms_representative_protein\",\n",
    "        \"dmel_unique_protein_isoforms_FBgn\", \"dmel_unique_protein_isoforms_identical_protein(s)\", \"dmel_unique_protein_isoforms_FB_gene_symbol\",\n",
    "        \"dmel_gene_sequence_ontology_annotations_so_term_id\", \"dmel_gene_sequence_ontology_annotations_gene_symbol\",\n",
    "        \"dmel_gene_sequence_ontology_annotations_gene_primary_id\", \"scRNA-Seq_gene_expression_Source_Tissue_Sex\",\n",
    "        \"scRNA-Seq_gene_expression_Clustering_Analysis_Name\", \"scRNA-Seq_gene_expression_Cluster_Cell_Type_Name\",\n",
    "        \"scRNA-Seq_gene_expression_Pub_miniref\", \"scRNA-Seq_gene_expression_Source_Tissue_Stage\", \"scRNA-Seq_gene_expression_Clustering_Analysis_ID\",\n",
    "        \"scRNA-Seq_gene_expression_Gene_Symbol\", \"scRNA-Seq_gene_expression_Spread\", \"scRNA-Seq_gene_expression_Pub_ID\",\n",
    "        \"scRNA-Seq_gene_expression_Gene_ID\", \"scRNA-Seq_gene_expression_Source_Tissue_Anatomy\", \"scRNA-Seq_gene_expression_Mean_Expression\",\n",
    "        \"scRNA-Seq_gene_expression_Cluster_ID\", \"scRNA-Seq_gene_expression_Cluster_Name\", \"scRNA-Seq_gene_expression_Cluster_Cell_Type_ID\",\n",
    "        \"gene_genetic_interactions_Interaction_type\", \"gene_genetic_interactions_Interacting_gene(s)_symbol\",\n",
    "        \"gene_genetic_interactions_Interacting_gene(s)_FBgn\", \"gene_genetic_interactions_Starting_gene(s)_FBgn\",\n",
    "        \"gene_genetic_interactions_Publication_FBrf\", \"gene_genetic_interactions_Starting_gene(s)_symbol\",\n",
    "        \"gene_rpkm_report_Bin_value\", \"gene_rpkm_report_Parent_library_FBlc#\", \"gene_rpkm_report_Total_exon_base_count\",\n",
    "        \"gene_rpkm_report_Parent_library_name\", \"gene_rpkm_report_RNASource_name\", \"gene_rpkm_report_FBgn#\",\n",
    "        \"gene_rpkm_report_RNASource_FBlc#\", \"gene_rpkm_report_GeneSymbol\", \"gene_rpkm_report_Count_used\",\n",
    "        \"gene_rpkm_report_Unique_exon_base_count\", \"gene_rpkm_report_Release_ID\", \"gene_rpkm_report_RPKM_value\",\n",
    "        \"physical_interactions_mitab_Interaction_Identifier(s)\", \"physical_interactions_mitab_Interaction_Type(s)\",\n",
    "        \"physical_interactions_mitab_Annotation(s)_Interactor_B\", \"physical_interactions_mitab_Publication_1st_Author(s)\",\n",
    "        \"physical_interactions_mitab_Alias(es)_Interactor_B\", \"physical_interactions_mitab_Alt_ID(s)_Interactor_A\",\n",
    "        \"physical_interactions_mitab_Alt_ID(s)_Interactor_B\", \"physical_interactions_mitab_Interaction_Xref(s)\",\n",
    "        \"physical_interactions_mitab_Experimental_Role(s)_Interactor_B\", \"physical_interactions_mitab_ID(s)_Interactor_A\",\n",
    "        \"physical_interactions_mitab_Type(s)_Interactor_A\", \"physical_interactions_mitab_Biological_Role(s)_Interactor_A\",\n",
    "        \"physical_interactions_mitab_Publication_ID(s)\", \"physical_interactions_mitab_ID(s)_Interactor_B\",\n",
    "        \"physical_interactions_mitab_Experimental_Role(s)_Interactor_A\", \"physical_interactions_mitab_Type(s)_Interactor_B\",\n",
    "        \"physical_interactions_mitab_Xref(s)_Interactor_B\", \"physical_interactions_mitab_Host_Organism(s)\",\n",
    "        \"physical_interactions_mitab_Biological_Role(s)_Interactor_B\", \"physical_interactions_mitab_Interaction_Detection_Method(s)\",\n",
    "        \"physical_interactions_mitab_Annotation(s)_Interactor_A\", \"physical_interactions_mitab_Source_Database(s)\",\n",
    "        \"physical_interactions_mitab_Taxid_Interactor_B\", \"physical_interactions_mitab_Taxid_Interactor_A\",\n",
    "        \"physical_interactions_mitab_Xref(s)_Interactor_A\", \"physical_interactions_mitab_Interaction_Annotation(s)\",\n",
    "        \"physical_interactions_mitab_Alias(es)_Interactor_A\", \"fb_synonym_symbol_synonym(s)\", \"fb_synonym_organism_abbreviation\",\n",
    "        \"fb_synonym_fullname_synonym(s)\", \"fb_synonym_primary_FBid\", \"fb_synonym_current_symbol\", \"fb_synonym_current_fullname\",\n",
    "        \"allele_genetic_interactions_allele_symbol\", \"allele_genetic_interactions_interaction\", \"allele_genetic_interactions_FBrf#\",\n",
    "        \"allele_genetic_interactions_allele_FBal#\", \"fbrf_pmid_pmcid_doi_DOI\", \"fbrf_pmid_pmcid_doi_PMCID\",\n",
    "        \"fbrf_pmid_pmcid_doi_pub_type\", \"fbrf_pmid_pmcid_doi_miniref\", \"fbrf_pmid_pmcid_doi_FBrf\",\n",
    "        \"fbrf_pmid_pmcid_doi_pmid_added\", \"fbrf_pmid_pmcid_doi_PMID\", \"disease_model_annotations_Based_on_orthology_with_(HGNC_ID)\",\n",
    "        \"disease_model_annotations_Evidence/interacting_alleles\", \"disease_model_annotations_Based_on_orthology_with_(symbol)\",\n",
    "        \"disease_model_annotations_DO_qualifier\", \"disease_model_annotations_DO_ID\", \"disease_model_annotations_Gene_symbol\",\n",
    "        \"disease_model_annotations_Allele_used_in_model_(FBal_ID)\", \"disease_model_annotations_Reference_(FBrf_ID)\",\n",
    "        \"disease_model_annotations_HGNC_ID\", \"disease_model_annotations_FBgn_ID\", \"disease_model_annotations_Allele_used_in_model_(symbol)\",\n",
    "        \"fbal_to_fbgn_AlleleID\", \"fbal_to_fbgn_GeneSymbol\", \"fbal_to_fbgn_AlleleSymbol\", \"fbal_to_fbgn_GeneID\",\n",
    "        \"gene_association_DB_Object_Name\", \"gene_association_With_(or)_From\", \"gene_association_GO_ID\",\n",
    "        \"gene_association_DB_Object_Type\", \"gene_association_DB_Object_Symbol\", \"gene_association_Evidence\",\n",
    "        \"gene_association_DB_Object_Synonym\", \"gene_association_DB_Object_ID\", \"gene_association_DB:Reference\",\n",
    "        \"gene_association_DB\", \"gene_association_Qualifier\", \"gene_association_Assigned_by\",\n",
    "        \"featureprop_value\",\n",
    "        \"featureprop_rank\",\n",
    "        \"pubprop_value\",\n",
    "        \"pubprop_rank\",\n",
    "        \"allele_symbol\",\n",
    "        \"allele_is_alleleof\",\n",
    "        \"allele_is_construct\",\n",
    "        \"allele_propagate_transgenic_uses\",\n",
    "        \"allele_gene_is_regulatory_region\",\n",
    "        \"allele_stocks_count\",\n",
    "        \"allele_pub_count\",\n",
    "        \"allele_known_lesion\",\n",
    "        \"allele_has_image\",\n",
    "        \"feature_name\",\n",
    "        \"feature_uniquename\",\n",
    "        \"feature_residues\",\n",
    "        \"feature_seqlen\",\n",
    "        \"grp_name\",\n",
    "        \"grp_uniquename\",\n",
    "        \"grp_synonym_grp_id\",\n",
    "        \"library_name\",\n",
    "        \"library_uniquename\",\n",
    "        \"feature_synonym_feature_id\",\n",
    "        \"feature_is_analysis\",\n",
    "        \"feature_is_obsolete\",\n",
    "        \"featureloc_fmin\",\n",
    "        \"featureloc_is_fmin_partial\",\n",
    "        \"featureloc_fmax\",\n",
    "        \"featureloc_is_fmax_partial\",\n",
    "        \"featureloc_strand\",\n",
    "        \"featureloc_phase\",\n",
    "        \"featureloc_residue_info\",\n",
    "        \"featureloc_locgroup\",\n",
    "        \"featureloc_rank\",\n",
    "        \"gene_name\",\n",
    "        \"gene_uniquename\",\n",
    "        \"gene_residues\",\n",
    "        \"gene_seqlen\",\n",
    "        \"gene_md5checksum\",\n",
    "        \"gene_is_analysis\",\n",
    "        \"gene_is_obsolete\",\n",
    "        \"cvterm_definition\",\n",
    "        \"cvterm_is_obsolete\",\n",
    "        \"cvterm_is_relationshiptype\",\n",
    "        \"dbxref_accession\",\n",
    "        \"dbxref_version\",\n",
    "        \"dbxref_description\",\n",
    "        \"dbxref_url\",\n",
    "        \"organism_abbreviation\",\n",
    "        \"organism_genus\",\n",
    "        \"organism_species\",\n",
    "        \"organism_common_name\",\n",
    "        \"organism_comment\",\n",
    "        \"pub_title\",\n",
    "        \"pub_volumetitle\",\n",
    "        \"pub_volume\",\n",
    "        \"pub_series_name\",\n",
    "        \"pub_issue\",\n",
    "        \"pub_pyear\",\n",
    "        \"pub_pages\",\n",
    "        \"pub_miniref\",\n",
    "        \"pub_is_obsolete\",\n",
    "        \"pub_publisher\",\n",
    "        \"pub_pubplace\",\n",
    "        \"synonym_name\",\n",
    "        \"synonym_synonym_sgml\"\n",
    "        ]\n",
    "\n",
    "\n",
    "        for column in schema_names:\n",
    "            v1 = Variable(\"v1\")\n",
    "            v2 = Variable(\"v2\")\n",
    "            s =  Node(\"Schema\", \"Schema:\" + column)\n",
    "            # s1 = Node(\"Schema\", \"Schema:gene_residues\")\n",
    "            q1 = Link(\"Execution\", ordered=True, targets=[s, v1, v2])  # linka schema=s à v1 e v2. v1 = pk e v2 = column\n",
    "            # q2 = Link(\"Execution\", ordered=True, targets=[s1, v1, v2])\n",
    "            assignments = query(q1, True)\n",
    "            # ass2 = query(q2, True)\n",
    "\n",
    "            cont = 0\n",
    "            print(f\"Number of relationships in column {column}: {len(assignments)}\")\n",
    "            if print_full_data:\n",
    "                for assignment in assignments:\n",
    "                    # pkey_handle = assignment.mapping[\"v1\"]\n",
    "                    # pkey = db.get_node_name(pkey_handle)\n",
    "                    un_handle = assignment.mapping[\"v2\"]\n",
    "                    unique_name = db.get_node_name(un_handle)\n",
    "                    print(str(cont) + \": \" + unique_name)\n",
    "                    cont += 1\n",
    "                    #if cont > 100:\n",
    "                        #break\n",
    "                    # if unique_name == verificar_prefixo(unique_name):\n",
    "                    #    print(\"v2: \" + unique_name)\n",
    "            print(\"FINISHED for column: \" + column + \"\\n\")\n",
    "\n",
    "\n",
    "    def print_schema_nodes_names(self, file_absolute_name, print_full_data=False):\n",
    "        schema_names = self.read_precomputed_table_columns(file_absolute_name)\n",
    "\n",
    "        schema_nodes_names = \"schema_nodes_names = [\"\n",
    "        for i in range(len(schema_names)):\n",
    "            if i != ( len(schema_names)  - 1 ):\n",
    "                schema_nodes_names += f'\"{schema_names[i]}\",'\n",
    "            else:\n",
    "                schema_nodes_names += f'\"{schema_names[i]}\"]'\n",
    "        print(schema_nodes_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bbddfe0c-f5a2-4f21-8993-be38d7c90ef1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54 milliseconds\n",
      "True\n",
      "8359 answers\n",
      "Number of relationships in column ncRNA_genes_soTermId: 8359\n",
      "FINISHED for column: ncRNA_genes_soTermId\n",
      "\n",
      "121 milliseconds\n",
      "True\n",
      "18144 answers\n",
      "Number of relationships in column ncRNA_genes_primaryId: 18144\n",
      "FINISHED for column: ncRNA_genes_primaryId\n",
      "\n",
      "21 milliseconds\n",
      "True\n",
      "3547 answers\n",
      "Number of relationships in column ncRNA_genes_gene_symbol: 3547\n",
      "FINISHED for column: ncRNA_genes_gene_symbol\n",
      "\n",
      "114 milliseconds\n",
      "True\n",
      "18986 answers\n",
      "Number of relationships in column ncRNA_genes_gene_geneId: 18986\n",
      "FINISHED for column: ncRNA_genes_gene_geneId\n",
      "\n",
      "27 milliseconds\n",
      "True\n",
      "4536 answers\n",
      "Number of relationships in column ncRNA_genes_symbol: 4536\n",
      "FINISHED for column: ncRNA_genes_symbol\n",
      "\n",
      "21 milliseconds\n",
      "True\n",
      "3547 answers\n",
      "Number of relationships in column ncRNA_genes_gene_locusTag: 3547\n",
      "FINISHED for column: ncRNA_genes_gene_locusTag\n",
      "\n",
      "47 milliseconds\n",
      "True\n",
      "8097 answers\n",
      "Number of relationships in column ncRNA_genes_taxonId: 8097\n",
      "FINISHED for column: ncRNA_genes_taxonId\n",
      "\n",
      "55 milliseconds\n",
      "True\n",
      "9194 answers\n",
      "Number of relationships in column ncRNA_genes_synonyms_symbol2: 9194\n",
      "FINISHED for column: ncRNA_genes_synonyms_symbol2\n",
      "\n",
      "56 milliseconds\n",
      "True\n",
      "9194 answers\n",
      "Number of relationships in column ncRNA_genes_synonyms_symbol1: 9194\n",
      "FINISHED for column: ncRNA_genes_synonyms_symbol1\n",
      "\n",
      "56 milliseconds\n",
      "True\n",
      "9520 answers\n",
      "Number of relationships in column ncRNA_genes_cross_references_symbol2: 9520\n",
      "FINISHED for column: ncRNA_genes_cross_references_symbol2\n",
      "\n",
      "56 milliseconds\n",
      "True\n",
      "9520 answers\n",
      "Number of relationships in column ncRNA_genes_cross_references_symbol1: 9520\n",
      "FINISHED for column: ncRNA_genes_cross_references_symbol1\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1714 answers\n",
      "Number of relationships in column ncRNA_genes_related_sequences_sequenceId: 1714\n",
      "FINISHED for column: ncRNA_genes_related_sequences_sequenceId\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "746 answers\n",
      "Number of relationships in column ncRNA_genes_related_sequences_relationship: 746\n",
      "FINISHED for column: ncRNA_genes_related_sequences_relationship\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "968 answers\n",
      "Number of relationships in column ncRNA_genes_related_sequences_primaryId: 968\n",
      "FINISHED for column: ncRNA_genes_related_sequences_primaryId\n",
      "\n",
      "682 milliseconds\n",
      "True\n",
      "20125 answers\n",
      "Number of relationships in column ncRNA_genes_gene_synonyms_symbol2: 20125\n",
      "FINISHED for column: ncRNA_genes_gene_synonyms_symbol2\n",
      "\n",
      "122 milliseconds\n",
      "True\n",
      "20125 answers\n",
      "Number of relationships in column ncRNA_genes_gene_synonyms_symbol1: 20125\n",
      "FINISHED for column: ncRNA_genes_gene_synonyms_symbol1\n",
      "\n",
      "28 milliseconds\n",
      "True\n",
      "4696 answers\n",
      "Number of relationships in column ncRNA_genes_publications_publication: 4696\n",
      "FINISHED for column: ncRNA_genes_publications_publication\n",
      "\n",
      "29 milliseconds\n",
      "True\n",
      "4696 answers\n",
      "Number of relationships in column ncRNA_genes_publications_primaryId: 4696\n",
      "FINISHED for column: ncRNA_genes_publications_primaryId\n",
      "\n",
      "80 milliseconds\n",
      "True\n",
      "13608 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_primaryId: 13608\n",
      "FINISHED for column: ncRNA_genes_genome_locations_primaryId\n",
      "\n",
      "67 milliseconds\n",
      "True\n",
      "11294 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_endPosition: 11294\n",
      "FINISHED for column: ncRNA_genes_genome_locations_endPosition\n",
      "\n",
      "96 milliseconds\n",
      "True\n",
      "15883 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_assembly: 15883\n",
      "FINISHED for column: ncRNA_genes_genome_locations_assembly\n",
      "\n",
      "29 milliseconds\n",
      "True\n",
      "4561 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_gca_accession: 4561\n",
      "FINISHED for column: ncRNA_genes_genome_locations_gca_accession\n",
      "\n",
      "95 milliseconds\n",
      "True\n",
      "15956 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_INSDC_accession: 15956\n",
      "FINISHED for column: ncRNA_genes_genome_locations_INSDC_accession\n",
      "\n",
      "86 milliseconds\n",
      "True\n",
      "11327 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_startPosition: 11327\n",
      "FINISHED for column: ncRNA_genes_genome_locations_startPosition\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "40 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_strand: 40\n",
      "FINISHED for column: ncRNA_genes_genome_locations_strand\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "41 answers\n",
      "Number of relationships in column ncRNA_genes_genome_locations_chromosome: 41\n",
      "FINISHED for column: ncRNA_genes_genome_locations_chromosome\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "23 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_gene_symbol: 23\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_gene_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "80 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_polypeptide_symbol: 80\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_polypeptide_symbol\n",
      "\n",
      "3 milliseconds\n",
      "True\n",
      "400 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_transcript_ID: 400\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_transcript_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "80 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_transcript_symbol: 80\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_transcript_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "80 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_transcript_type: 80\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_transcript_type\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "183 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_annotation_ID: 183\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_annotation_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "23 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_organism: 23\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_organism\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "320 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_polypeptide_ID: 320\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_polypeptide_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "23 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_gene_fullname: 23\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_gene_fullname\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "250 answers\n",
      "Number of relationships in column fbgn_fbtr_fbpp_expanded_gene_ID: 250\n",
      "FINISHED for column: fbgn_fbtr_fbpp_expanded_gene_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "19 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_Dmel_gene_symbol: 19\n",
      "FINISHED for column: dmel_human_orthologs_disease_Dmel_gene_symbol\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "268 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_Human_gene_HGNC_ID: 268\n",
      "FINISHED for column: dmel_human_orthologs_disease_Human_gene_HGNC_ID\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "262 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_Human_gene_OMIM_ID: 262\n",
      "FINISHED for column: dmel_human_orthologs_disease_Human_gene_OMIM_ID\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "209 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_Dmel_gene_ID: 209\n",
      "FINISHED for column: dmel_human_orthologs_disease_Dmel_gene_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "120 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_Human_gene_symbol: 120\n",
      "FINISHED for column: dmel_human_orthologs_disease_Human_gene_symbol\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "190 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_DIOPT_score: 190\n",
      "FINISHED for column: dmel_human_orthologs_disease_DIOPT_score\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "84 answers\n",
      "Number of relationships in column dmel_human_orthologs_disease_OMIM_Phenotype_IDs[name]: 84\n",
      "FINISHED for column: dmel_human_orthologs_disease_OMIM_Phenotype_IDs[name]\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "5 answers\n",
      "Number of relationships in column Dmel_enzyme_data_gene_symbol: 5\n",
      "FINISHED for column: Dmel_enzyme_data_gene_symbol\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column Dmel_enzyme_data_gene_EC_number(s): 0\n",
      "FINISHED for column: Dmel_enzyme_data_gene_EC_number(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column Dmel_enzyme_data_gene_group_EC_number(s): 0\n",
      "FINISHED for column: Dmel_enzyme_data_gene_group_EC_number(s)\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "5 answers\n",
      "Number of relationships in column Dmel_enzyme_data_gene_name: 5\n",
      "FINISHED for column: Dmel_enzyme_data_gene_name\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "5 answers\n",
      "Number of relationships in column Dmel_enzyme_data_gene_group_name: 5\n",
      "FINISHED for column: Dmel_enzyme_data_gene_group_name\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column Dmel_enzyme_data_gene_group_GO_id(s): 0\n",
      "FINISHED for column: Dmel_enzyme_data_gene_group_GO_id(s)\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "24 answers\n",
      "Number of relationships in column Dmel_enzyme_data_gene_id: 24\n",
      "FINISHED for column: Dmel_enzyme_data_gene_id\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "19 answers\n",
      "Number of relationships in column Dmel_enzyme_data_gene_group_id: 19\n",
      "FINISHED for column: Dmel_enzyme_data_gene_group_id\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1702 answers\n",
      "Number of relationships in column gene_groups_HGNC_FB_group_name: 1702\n",
      "FINISHED for column: gene_groups_HGNC_FB_group_name\n",
      "\n",
      "10 milliseconds\n",
      "True\n",
      "1702 answers\n",
      "Number of relationships in column gene_groups_HGNC_FB_group_symbol: 1702\n",
      "FINISHED for column: gene_groups_HGNC_FB_group_symbol\n",
      "\n",
      "23 milliseconds\n",
      "True\n",
      "3923 answers\n",
      "Number of relationships in column gene_groups_HGNC_FB_group_id: 3923\n",
      "FINISHED for column: gene_groups_HGNC_FB_group_id\n",
      "\n",
      "4 milliseconds\n",
      "True\n",
      "519 answers\n",
      "Number of relationships in column gene_groups_HGNC_HGNC_family_ID: 519\n",
      "FINISHED for column: gene_groups_HGNC_HGNC_family_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "92 answers\n",
      "Number of relationships in column best_gene_summary_Summary: 92\n",
      "FINISHED for column: best_gene_summary_Summary\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "92 answers\n",
      "Number of relationships in column best_gene_summary_Summary_Source: 92\n",
      "FINISHED for column: best_gene_summary_Summary_Source\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "92 answers\n",
      "Number of relationships in column best_gene_summary_Gene_Symbol: 92\n",
      "FINISHED for column: best_gene_summary_Gene_Symbol\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "276 answers\n",
      "Number of relationships in column best_gene_summary_FBgn_ID: 276\n",
      "FINISHED for column: best_gene_summary_FBgn_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "23 answers\n",
      "Number of relationships in column gene_map_table_cytogenetic_loc: 23\n",
      "FINISHED for column: gene_map_table_cytogenetic_loc\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "80 answers\n",
      "Number of relationships in column gene_map_table_organism_abbreviation: 80\n",
      "FINISHED for column: gene_map_table_organism_abbreviation\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "25 answers\n",
      "Number of relationships in column gene_map_table_recombination_loc: 25\n",
      "FINISHED for column: gene_map_table_recombination_loc\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "229 answers\n",
      "Number of relationships in column gene_map_table_primary_FBid: 229\n",
      "FINISHED for column: gene_map_table_primary_FBid\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "80 answers\n",
      "Number of relationships in column gene_map_table_current_symbol: 80\n",
      "FINISHED for column: gene_map_table_current_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column gene_map_table_sequence_loc: 21\n",
      "FINISHED for column: gene_map_table_sequence_loc\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "17 answers\n",
      "Number of relationships in column pathway_group_data_Parent_FB_group_symbol: 17\n",
      "FINISHED for column: pathway_group_data_Parent_FB_group_symbol\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "1008 answers\n",
      "Number of relationships in column pathway_group_data_Parent_FB_group_id: 1008\n",
      "FINISHED for column: pathway_group_data_Parent_FB_group_id\n",
      "\n",
      "15 milliseconds\n",
      "True\n",
      "2570 answers\n",
      "Number of relationships in column pathway_group_data_Group_member_FB_gene_id: 2570\n",
      "FINISHED for column: pathway_group_data_Group_member_FB_gene_id\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "71 answers\n",
      "Number of relationships in column pathway_group_data_FB_group_name: 71\n",
      "FINISHED for column: pathway_group_data_FB_group_name\n",
      "\n",
      "4 milliseconds\n",
      "True\n",
      "657 answers\n",
      "Number of relationships in column pathway_group_data_Group_member_FB_gene_symbol: 657\n",
      "FINISHED for column: pathway_group_data_Group_member_FB_gene_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "71 answers\n",
      "Number of relationships in column pathway_group_data_FB_group_symbol: 71\n",
      "FINISHED for column: pathway_group_data_FB_group_symbol\n",
      "\n",
      "7 milliseconds\n",
      "True\n",
      "1172 answers\n",
      "Number of relationships in column pathway_group_data_FB_group_id: 1172\n",
      "FINISHED for column: pathway_group_data_FB_group_id\n",
      "\n",
      "25 milliseconds\n",
      "True\n",
      "4343 answers\n",
      "Number of relationships in column genotype_phenotype_data_qualifier_ids: 4343\n",
      "FINISHED for column: genotype_phenotype_data_qualifier_ids\n",
      "\n",
      "1.236 seconds\n",
      "True\n",
      "101628 answers\n",
      "Number of relationships in column genotype_phenotype_data_genotype_symbols: 101628\n",
      "FINISHED for column: genotype_phenotype_data_genotype_symbols\n",
      "\n",
      "3 milliseconds\n",
      "True\n",
      "229 answers\n",
      "Number of relationships in column genotype_phenotype_data_qualifier_names: 229\n",
      "FINISHED for column: genotype_phenotype_data_qualifier_names\n",
      "\n",
      "218 milliseconds\n",
      "True\n",
      "36146 answers\n",
      "Number of relationships in column genotype_phenotype_data_phenotype_id: 36146\n",
      "FINISHED for column: genotype_phenotype_data_phenotype_id\n",
      "\n",
      "1.681 seconds\n",
      "True\n",
      "166217 answers\n",
      "Number of relationships in column genotype_phenotype_data_genotype_FBids: 166217\n",
      "FINISHED for column: genotype_phenotype_data_genotype_FBids\n",
      "\n",
      "5 milliseconds\n",
      "True\n",
      "680 answers\n",
      "Number of relationships in column genotype_phenotype_data_phenotype_name: 680\n",
      "FINISHED for column: genotype_phenotype_data_phenotype_name\n",
      "\n",
      "164 milliseconds\n",
      "True\n",
      "26843 answers\n",
      "Number of relationships in column genotype_phenotype_data_reference: 26843\n",
      "FINISHED for column: genotype_phenotype_data_reference\n",
      "\n",
      "3 milliseconds\n",
      "True\n",
      "388 answers\n",
      "Number of relationships in column gene_group_data_Parent_FB_group_symbol: 388\n",
      "FINISHED for column: gene_group_data_Parent_FB_group_symbol\n",
      "\n",
      "68 milliseconds\n",
      "True\n",
      "10987 answers\n",
      "Number of relationships in column gene_group_data_Parent_FB_group_id: 10987\n",
      "FINISHED for column: gene_group_data_Parent_FB_group_id\n",
      "\n",
      "168 milliseconds\n",
      "True\n",
      "27122 answers\n",
      "Number of relationships in column gene_group_data_Group_member_FB_gene_id: 27122\n",
      "FINISHED for column: gene_group_data_Group_member_FB_gene_id\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1631 answers\n",
      "Number of relationships in column gene_group_data_FB_group_name: 1631\n",
      "FINISHED for column: gene_group_data_FB_group_name\n",
      "\n",
      "50 milliseconds\n",
      "True\n",
      "8253 answers\n",
      "Number of relationships in column gene_group_data_Group_member_FB_gene_symbol: 8253\n",
      "FINISHED for column: gene_group_data_Group_member_FB_gene_symbol\n",
      "\n",
      "10 milliseconds\n",
      "True\n",
      "1631 answers\n",
      "Number of relationships in column gene_group_data_FB_group_symbol: 1631\n",
      "FINISHED for column: gene_group_data_FB_group_symbol\n",
      "\n",
      "86 milliseconds\n",
      "True\n",
      "14614 answers\n",
      "Number of relationships in column gene_group_data_FB_group_id: 14614\n",
      "FINISHED for column: gene_group_data_FB_group_id\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "53 answers\n",
      "Number of relationships in column dmel_unique_protein_isoforms_representative_protein: 53\n",
      "FINISHED for column: dmel_unique_protein_isoforms_representative_protein\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "98 answers\n",
      "Number of relationships in column dmel_unique_protein_isoforms_FBgn: 98\n",
      "FINISHED for column: dmel_unique_protein_isoforms_FBgn\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column dmel_unique_protein_isoforms_identical_protein(s): 0\n",
      "FINISHED for column: dmel_unique_protein_isoforms_identical_protein(s)\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column dmel_unique_protein_isoforms_FB_gene_symbol: 21\n",
      "FINISHED for column: dmel_unique_protein_isoforms_FB_gene_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "49 answers\n",
      "Number of relationships in column dmel_gene_sequence_ontology_annotations_so_term_id: 49\n",
      "FINISHED for column: dmel_gene_sequence_ontology_annotations_so_term_id\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column dmel_gene_sequence_ontology_annotations_gene_symbol: 21\n",
      "FINISHED for column: dmel_gene_sequence_ontology_annotations_gene_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "70 answers\n",
      "Number of relationships in column dmel_gene_sequence_ontology_annotations_gene_primary_id: 70\n",
      "FINISHED for column: dmel_gene_sequence_ontology_annotations_gene_primary_id\n",
      "\n",
      "10 milliseconds\n",
      "True\n",
      "1434 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Source_Tissue_Sex: 1434\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Source_Tissue_Sex\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "89 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Clustering_Analysis_Name: 89\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Clustering_Analysis_Name\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "271 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Cluster_Cell_Type_Name: 271\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Cluster_Cell_Type_Name\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "13 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Pub_miniref: 13\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Pub_miniref\n",
      "\n",
      "12 milliseconds\n",
      "True\n",
      "2050 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Source_Tissue_Stage: 2050\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Source_Tissue_Stage\n",
      "\n",
      "33 milliseconds\n",
      "True\n",
      "5638 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Clustering_Analysis_ID: 5638\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Clustering_Analysis_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Gene_Symbol: 21\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Gene_Symbol\n",
      "\n",
      "109 milliseconds\n",
      "True\n",
      "18294 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Spread: 18294\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Spread\n",
      "\n",
      "15 milliseconds\n",
      "True\n",
      "2421 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Pub_ID: 2421\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Pub_ID\n",
      "\n",
      "461 milliseconds\n",
      "True\n",
      "76308 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Gene_ID: 76308\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Gene_ID\n",
      "\n",
      "15 milliseconds\n",
      "True\n",
      "2435 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Source_Tissue_Anatomy: 2435\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Source_Tissue_Anatomy\n",
      "\n",
      "689 milliseconds\n",
      "True\n",
      "20734 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Mean_Expression: 20734\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Mean_Expression\n",
      "\n",
      "259 milliseconds\n",
      "True\n",
      "41242 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Cluster_ID: 41242\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Cluster_ID\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1690 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Cluster_Name: 1690\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Cluster_Name\n",
      "\n",
      "62 milliseconds\n",
      "True\n",
      "10358 answers\n",
      "Number of relationships in column scRNA-Seq_gene_expression_Cluster_Cell_Type_ID: 10358\n",
      "FINISHED for column: scRNA-Seq_gene_expression_Cluster_Cell_Type_ID\n",
      "\n",
      "4 milliseconds\n",
      "True\n",
      "534 answers\n",
      "Number of relationships in column gene_genetic_interactions_Interaction_type: 534\n",
      "FINISHED for column: gene_genetic_interactions_Interaction_type\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_genetic_interactions_Interacting_gene(s)_symbol: 0\n",
      "FINISHED for column: gene_genetic_interactions_Interacting_gene(s)_symbol\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_genetic_interactions_Interacting_gene(s)_FBgn: 0\n",
      "FINISHED for column: gene_genetic_interactions_Interacting_gene(s)_FBgn\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_genetic_interactions_Starting_gene(s)_FBgn: 0\n",
      "FINISHED for column: gene_genetic_interactions_Starting_gene(s)_FBgn\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "956 answers\n",
      "Number of relationships in column gene_genetic_interactions_Publication_FBrf: 956\n",
      "FINISHED for column: gene_genetic_interactions_Publication_FBrf\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_genetic_interactions_Starting_gene(s)_symbol: 0\n",
      "FINISHED for column: gene_genetic_interactions_Starting_gene(s)_symbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "113 answers\n",
      "Number of relationships in column gene_rpkm_report_Bin_value: 113\n",
      "FINISHED for column: gene_rpkm_report_Bin_value\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "320 answers\n",
      "Number of relationships in column gene_rpkm_report_Parent_library_FBlc#: 320\n",
      "FINISHED for column: gene_rpkm_report_Parent_library_FBlc#\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column gene_rpkm_report_Total_exon_base_count: 21\n",
      "FINISHED for column: gene_rpkm_report_Total_exon_base_count\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "7 answers\n",
      "Number of relationships in column gene_rpkm_report_Parent_library_name: 7\n",
      "FINISHED for column: gene_rpkm_report_Parent_library_name\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "166 answers\n",
      "Number of relationships in column gene_rpkm_report_RNASource_name: 166\n",
      "FINISHED for column: gene_rpkm_report_RNASource_name\n",
      "\n",
      "28 milliseconds\n",
      "True\n",
      "4700 answers\n",
      "Number of relationships in column gene_rpkm_report_FBgn#: 4700\n",
      "FINISHED for column: gene_rpkm_report_FBgn#\n",
      "\n",
      "22 milliseconds\n",
      "True\n",
      "3818 answers\n",
      "Number of relationships in column gene_rpkm_report_RNASource_FBlc#: 3818\n",
      "FINISHED for column: gene_rpkm_report_RNASource_FBlc#\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column gene_rpkm_report_GeneSymbol: 21\n",
      "FINISHED for column: gene_rpkm_report_GeneSymbol\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column gene_rpkm_report_Count_used: 21\n",
      "FINISHED for column: gene_rpkm_report_Count_used\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "28 answers\n",
      "Number of relationships in column gene_rpkm_report_Unique_exon_base_count: 28\n",
      "FINISHED for column: gene_rpkm_report_Unique_exon_base_count\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "21 answers\n",
      "Number of relationships in column gene_rpkm_report_Release_ID: 21\n",
      "FINISHED for column: gene_rpkm_report_Release_ID\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "961 answers\n",
      "Number of relationships in column gene_rpkm_report_RPKM_value: 961\n",
      "FINISHED for column: gene_rpkm_report_RPKM_value\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Interaction_Identifier(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Interaction_Identifier(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Interaction_Type(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Interaction_Type(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Annotation(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Annotation(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Publication_1st_Author(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Publication_1st_Author(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Alias(es)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Alias(es)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Alt_ID(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Alt_ID(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Alt_ID(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Alt_ID(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Interaction_Xref(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Interaction_Xref(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Experimental_Role(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Experimental_Role(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_ID(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_ID(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Type(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Type(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Biological_Role(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Biological_Role(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Publication_ID(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Publication_ID(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_ID(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_ID(s)_Interactor_B\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Experimental_Role(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Experimental_Role(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Type(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Type(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Xref(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Xref(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Host_Organism(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Host_Organism(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Biological_Role(s)_Interactor_B: 0\n",
      "FINISHED for column: physical_interactions_mitab_Biological_Role(s)_Interactor_B\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Interaction_Detection_Method(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Interaction_Detection_Method(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Annotation(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Annotation(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Source_Database(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Source_Database(s)\n",
      "\n",
      "7 milliseconds\n",
      "True\n",
      "1102 answers\n",
      "Number of relationships in column physical_interactions_mitab_Taxid_Interactor_B: 1102\n",
      "FINISHED for column: physical_interactions_mitab_Taxid_Interactor_B\n",
      "\n",
      "5 milliseconds\n",
      "True\n",
      "849 answers\n",
      "Number of relationships in column physical_interactions_mitab_Taxid_Interactor_A: 849\n",
      "FINISHED for column: physical_interactions_mitab_Taxid_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Xref(s)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Xref(s)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Interaction_Annotation(s): 0\n",
      "FINISHED for column: physical_interactions_mitab_Interaction_Annotation(s)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column physical_interactions_mitab_Alias(es)_Interactor_A: 0\n",
      "FINISHED for column: physical_interactions_mitab_Alias(es)_Interactor_A\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column fb_synonym_symbol_synonym(s): 0\n",
      "FINISHED for column: fb_synonym_symbol_synonym(s)\n",
      "\n",
      "26 milliseconds\n",
      "True\n",
      "4554 answers\n",
      "Number of relationships in column fb_synonym_organism_abbreviation: 4554\n",
      "FINISHED for column: fb_synonym_organism_abbreviation\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column fb_synonym_fullname_synonym(s): 0\n",
      "FINISHED for column: fb_synonym_fullname_synonym(s)\n",
      "\n",
      "101 milliseconds\n",
      "True\n",
      "17147 answers\n",
      "Number of relationships in column fb_synonym_primary_FBid: 17147\n",
      "FINISHED for column: fb_synonym_primary_FBid\n",
      "\n",
      "28 milliseconds\n",
      "True\n",
      "4554 answers\n",
      "Number of relationships in column fb_synonym_current_symbol: 4554\n",
      "FINISHED for column: fb_synonym_current_symbol\n",
      "\n",
      "2 milliseconds\n",
      "True\n",
      "168 answers\n",
      "Number of relationships in column fb_synonym_current_fullname: 168\n",
      "FINISHED for column: fb_synonym_current_fullname\n",
      "\n",
      "12 milliseconds\n",
      "True\n",
      "1957 answers\n",
      "Number of relationships in column allele_genetic_interactions_allele_symbol: 1957\n",
      "FINISHED for column: allele_genetic_interactions_allele_symbol\n",
      "\n",
      "146 milliseconds\n",
      "True\n",
      "24509 answers\n",
      "Number of relationships in column allele_genetic_interactions_interaction: 24509\n",
      "FINISHED for column: allele_genetic_interactions_interaction\n",
      "\n",
      "72 milliseconds\n",
      "True\n",
      "11898 answers\n",
      "Number of relationships in column allele_genetic_interactions_FBrf#: 11898\n",
      "FINISHED for column: allele_genetic_interactions_FBrf#\n",
      "\n",
      "141 milliseconds\n",
      "True\n",
      "20392 answers\n",
      "Number of relationships in column allele_genetic_interactions_allele_FBal#: 20392\n",
      "FINISHED for column: allele_genetic_interactions_allele_FBal#\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "67 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_DOI: 67\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_DOI\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "25 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_PMCID: 25\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_PMCID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "75 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_pub_type: 75\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_pub_type\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "75 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_miniref: 75\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_miniref\n",
      "\n",
      "3 milliseconds\n",
      "True\n",
      "392 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_FBrf: 392\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_FBrf\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "75 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_pmid_added: 75\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_pmid_added\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "75 answers\n",
      "Number of relationships in column fbrf_pmid_pmcid_doi_PMID: 75\n",
      "FINISHED for column: fbrf_pmid_pmcid_doi_PMID\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column disease_model_annotations_Based_on_orthology_with_(HGNC_ID): 0\n",
      "FINISHED for column: disease_model_annotations_Based_on_orthology_with_(HGNC_ID)\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "1026 answers\n",
      "Number of relationships in column disease_model_annotations_Evidence/interacting_alleles: 1026\n",
      "FINISHED for column: disease_model_annotations_Evidence/interacting_alleles\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column disease_model_annotations_Based_on_orthology_with_(symbol): 0\n",
      "FINISHED for column: disease_model_annotations_Based_on_orthology_with_(symbol)\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "88 answers\n",
      "Number of relationships in column disease_model_annotations_DO_qualifier: 88\n",
      "FINISHED for column: disease_model_annotations_DO_qualifier\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "933 answers\n",
      "Number of relationships in column disease_model_annotations_DO_ID: 933\n",
      "FINISHED for column: disease_model_annotations_DO_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "150 answers\n",
      "Number of relationships in column disease_model_annotations_Gene_symbol: 150\n",
      "FINISHED for column: disease_model_annotations_Gene_symbol\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column disease_model_annotations_Allele_used_in_model_(FBal_ID): 0\n",
      "FINISHED for column: disease_model_annotations_Allele_used_in_model_(FBal_ID)\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column disease_model_annotations_Reference_(FBrf_ID): 0\n",
      "FINISHED for column: disease_model_annotations_Reference_(FBrf_ID)\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "63 answers\n",
      "Number of relationships in column disease_model_annotations_HGNC_ID: 63\n",
      "FINISHED for column: disease_model_annotations_HGNC_ID\n",
      "\n",
      "7 milliseconds\n",
      "True\n",
      "1183 answers\n",
      "Number of relationships in column disease_model_annotations_FBgn_ID: 1183\n",
      "FINISHED for column: disease_model_annotations_FBgn_ID\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column disease_model_annotations_Allele_used_in_model_(symbol): 0\n",
      "FINISHED for column: disease_model_annotations_Allele_used_in_model_(symbol)\n",
      "\n",
      "20 milliseconds\n",
      "True\n",
      "3430 answers\n",
      "Number of relationships in column fbal_to_fbgn_AlleleID: 3430\n",
      "FINISHED for column: fbal_to_fbgn_AlleleID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "74 answers\n",
      "Number of relationships in column fbal_to_fbgn_GeneSymbol: 74\n",
      "FINISHED for column: fbal_to_fbgn_GeneSymbol\n",
      "\n",
      "10 milliseconds\n",
      "True\n",
      "1715 answers\n",
      "Number of relationships in column fbal_to_fbgn_AlleleSymbol: 1715\n",
      "FINISHED for column: fbal_to_fbgn_AlleleSymbol\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1789 answers\n",
      "Number of relationships in column fbal_to_fbgn_GeneID: 1789\n",
      "FINISHED for column: fbal_to_fbgn_GeneID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "35 answers\n",
      "Number of relationships in column gene_association_DB_Object_Name: 35\n",
      "FINISHED for column: gene_association_DB_Object_Name\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_association_With_(or)_From: 0\n",
      "FINISHED for column: gene_association_With_(or)_From\n",
      "\n",
      "14 milliseconds\n",
      "True\n",
      "2437 answers\n",
      "Number of relationships in column gene_association_GO_ID: 2437\n",
      "FINISHED for column: gene_association_GO_ID\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "35 answers\n",
      "Number of relationships in column gene_association_DB_Object_Type: 35\n",
      "FINISHED for column: gene_association_DB_Object_Type\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "35 answers\n",
      "Number of relationships in column gene_association_DB_Object_Symbol: 35\n",
      "FINISHED for column: gene_association_DB_Object_Symbol\n",
      "\n",
      "6 milliseconds\n",
      "True\n",
      "927 answers\n",
      "Number of relationships in column gene_association_Evidence: 927\n",
      "FINISHED for column: gene_association_Evidence\n",
      "\n",
      "5 milliseconds\n",
      "True\n",
      "881 answers\n",
      "Number of relationships in column gene_association_DB_Object_Synonym: 881\n",
      "FINISHED for column: gene_association_DB_Object_Synonym\n",
      "\n",
      "17 milliseconds\n",
      "True\n",
      "2998 answers\n",
      "Number of relationships in column gene_association_DB_Object_ID: 2998\n",
      "FINISHED for column: gene_association_DB_Object_ID\n",
      "\n",
      "5 milliseconds\n",
      "True\n",
      "680 answers\n",
      "Number of relationships in column gene_association_DB:Reference: 680\n",
      "FINISHED for column: gene_association_DB:Reference\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "35 answers\n",
      "Number of relationships in column gene_association_DB: 35\n",
      "FINISHED for column: gene_association_DB\n",
      "\n",
      "4 milliseconds\n",
      "True\n",
      "654 answers\n",
      "Number of relationships in column gene_association_Qualifier: 654\n",
      "FINISHED for column: gene_association_Qualifier\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "122 answers\n",
      "Number of relationships in column gene_association_Assigned_by: 122\n",
      "FINISHED for column: gene_association_Assigned_by\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureprop_value: 0\n",
      "FINISHED for column: featureprop_value\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureprop_rank: 0\n",
      "FINISHED for column: featureprop_rank\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pubprop_value: 0\n",
      "FINISHED for column: pubprop_value\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pubprop_rank: 0\n",
      "FINISHED for column: pubprop_rank\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_symbol: 0\n",
      "FINISHED for column: allele_symbol\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_is_alleleof: 0\n",
      "FINISHED for column: allele_is_alleleof\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_is_construct: 0\n",
      "FINISHED for column: allele_is_construct\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_propagate_transgenic_uses: 0\n",
      "FINISHED for column: allele_propagate_transgenic_uses\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_gene_is_regulatory_region: 0\n",
      "FINISHED for column: allele_gene_is_regulatory_region\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_stocks_count: 0\n",
      "FINISHED for column: allele_stocks_count\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_pub_count: 0\n",
      "FINISHED for column: allele_pub_count\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_known_lesion: 0\n",
      "FINISHED for column: allele_known_lesion\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column allele_has_image: 0\n",
      "FINISHED for column: allele_has_image\n",
      "\n",
      "7.543 seconds\n",
      "True\n",
      "736654 answers\n",
      "Number of relationships in column feature_name: 736654\n",
      "FINISHED for column: feature_name\n",
      "\n",
      "5.667 seconds\n",
      "True\n",
      "736654 answers\n",
      "Number of relationships in column feature_uniquename: 736654\n",
      "FINISHED for column: feature_uniquename\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column feature_residues: 0\n",
      "FINISHED for column: feature_residues\n",
      "\n",
      "1.150 seconds\n",
      "True\n",
      "88662 answers\n",
      "Number of relationships in column feature_seqlen: 88662\n",
      "FINISHED for column: feature_seqlen\n",
      "\n",
      "11 milliseconds\n",
      "True\n",
      "1704 answers\n",
      "Number of relationships in column grp_name: 1704\n",
      "FINISHED for column: grp_name\n",
      "\n",
      "12 milliseconds\n",
      "True\n",
      "1704 answers\n",
      "Number of relationships in column grp_uniquename: 1704\n",
      "FINISHED for column: grp_uniquename\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column grp_synonym_grp_id: 0\n",
      "FINISHED for column: grp_synonym_grp_id\n",
      "\n",
      "12 milliseconds\n",
      "True\n",
      "1952 answers\n",
      "Number of relationships in column library_name: 1952\n",
      "FINISHED for column: library_name\n",
      "\n",
      "12 milliseconds\n",
      "True\n",
      "1952 answers\n",
      "Number of relationships in column library_uniquename: 1952\n",
      "FINISHED for column: library_uniquename\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column feature_synonym_feature_id: 0\n",
      "FINISHED for column: feature_synonym_feature_id\n",
      "\n",
      "6.812 seconds\n",
      "True\n",
      "736654 answers\n",
      "Number of relationships in column feature_is_analysis: 736654\n",
      "FINISHED for column: feature_is_analysis\n",
      "\n",
      "7.030 seconds\n",
      "True\n",
      "736654 answers\n",
      "Number of relationships in column feature_is_obsolete: 736654\n",
      "FINISHED for column: feature_is_obsolete\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_fmin: 0\n",
      "FINISHED for column: featureloc_fmin\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_is_fmin_partial: 0\n",
      "FINISHED for column: featureloc_is_fmin_partial\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_fmax: 0\n",
      "FINISHED for column: featureloc_fmax\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_is_fmax_partial: 0\n",
      "FINISHED for column: featureloc_is_fmax_partial\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_strand: 0\n",
      "FINISHED for column: featureloc_strand\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_phase: 0\n",
      "FINISHED for column: featureloc_phase\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_residue_info: 0\n",
      "FINISHED for column: featureloc_residue_info\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_locgroup: 0\n",
      "FINISHED for column: featureloc_locgroup\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column featureloc_rank: 0\n",
      "FINISHED for column: featureloc_rank\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_name: 0\n",
      "FINISHED for column: gene_name\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_uniquename: 0\n",
      "FINISHED for column: gene_uniquename\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_residues: 0\n",
      "FINISHED for column: gene_residues\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_seqlen: 0\n",
      "FINISHED for column: gene_seqlen\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_md5checksum: 0\n",
      "FINISHED for column: gene_md5checksum\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_is_analysis: 0\n",
      "FINISHED for column: gene_is_analysis\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column gene_is_obsolete: 0\n",
      "FINISHED for column: gene_is_obsolete\n",
      "\n",
      "23 milliseconds\n",
      "True\n",
      "3719 answers\n",
      "Number of relationships in column cvterm_definition: 3719\n",
      "FINISHED for column: cvterm_definition\n",
      "\n",
      "44 milliseconds\n",
      "True\n",
      "7405 answers\n",
      "Number of relationships in column cvterm_is_obsolete: 7405\n",
      "FINISHED for column: cvterm_is_obsolete\n",
      "\n",
      "45 milliseconds\n",
      "True\n",
      "7405 answers\n",
      "Number of relationships in column cvterm_is_relationshiptype: 7405\n",
      "FINISHED for column: cvterm_is_relationshiptype\n",
      "\n",
      "4.592 seconds\n",
      "True\n",
      "621740 answers\n",
      "Number of relationships in column dbxref_accession: 621740\n",
      "FINISHED for column: dbxref_accession\n",
      "\n",
      "6.025 seconds\n",
      "True\n",
      "621740 answers\n",
      "Number of relationships in column dbxref_version: 621740\n",
      "FINISHED for column: dbxref_version\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column dbxref_description: 0\n",
      "FINISHED for column: dbxref_description\n",
      "\n",
      "1 milliseconds\n",
      "False\n",
      "Number of relationships in column dbxref_url: 0\n",
      "FINISHED for column: dbxref_url\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "14 answers\n",
      "Number of relationships in column organism_abbreviation: 14\n",
      "FINISHED for column: organism_abbreviation\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "16 answers\n",
      "Number of relationships in column organism_genus: 16\n",
      "FINISHED for column: organism_genus\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "16 answers\n",
      "Number of relationships in column organism_species: 16\n",
      "FINISHED for column: organism_species\n",
      "\n",
      "1 milliseconds\n",
      "True\n",
      "4 answers\n",
      "Number of relationships in column organism_common_name: 4\n",
      "FINISHED for column: organism_common_name\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column organism_comment: 0\n",
      "FINISHED for column: organism_comment\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_title: 0\n",
      "FINISHED for column: pub_title\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_volumetitle: 0\n",
      "FINISHED for column: pub_volumetitle\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_volume: 0\n",
      "FINISHED for column: pub_volume\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_series_name: 0\n",
      "FINISHED for column: pub_series_name\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_issue: 0\n",
      "FINISHED for column: pub_issue\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_pyear: 0\n",
      "FINISHED for column: pub_pyear\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_pages: 0\n",
      "FINISHED for column: pub_pages\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_miniref: 0\n",
      "FINISHED for column: pub_miniref\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_is_obsolete: 0\n",
      "FINISHED for column: pub_is_obsolete\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_publisher: 0\n",
      "FINISHED for column: pub_publisher\n",
      "\n",
      "0 milliseconds\n",
      "False\n",
      "Number of relationships in column pub_pubplace: 0\n",
      "FINISHED for column: pub_pubplace\n",
      "\n",
      "1.721 seconds\n",
      "True\n",
      "283529 answers\n",
      "Number of relationships in column synonym_name: 283529\n",
      "FINISHED for column: synonym_name\n",
      "\n",
      "2.570 seconds\n",
      "True\n",
      "283529 answers\n",
      "Number of relationships in column synonym_synonym_sgml: 283529\n",
      "FINISHED for column: synonym_synonym_sgml\n"
     ]
    }
   ],
   "source": [
    "\n",
    "verifyer = DASLoadingVerifier()\n",
    "file_absolute_name = \"/home/saulo/snet/hyperon/das/das/flybase2metta/fb_data/2023_05/precomputed/essential_pairs.txt\"\n",
    "file_absolute_name = \"/mnt/hdd_2/saulo/snet/hyperon/das/data/flybase/input/2023_05/precomputed/essential_pairs.txt\"\n",
    "#verifyer.print_schema_nodes_data(file_absolute_name, print_full_data=True)\n",
    "verifyer.print_schema_nodes_data(file_absolute_name, print_full_data=False)\n",
    "#verifyer.print_schema_nodes_names(file_absolute_name, print_full_data=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "530974f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mef2: ('FBgn0011656', [])\n",
      "Clk: ('FBgn0023076', [])\n",
      "Myc: ('FBgn0262656', [])\n",
      "Abd-B: ('FBgn0000015', [])\n"
     ]
    }
   ],
   "source": [
    "# get FB id and group name(s) of gene designated by gene_symbol  <==> DOESN'T GET CORRECT RESULTS IN \"essential pairing\" mode because\n",
    "# there is no link from FB_group_name to Group_member_FB_gene_id\n",
    "def get_gene_group_names(gene_symbol):\n",
    "    fb_id = get_feature_fb_id(gene_symbol)    \n",
    "    n1 = Node(\"Verbatim\", fb_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    # groups\n",
    "    s = Node(\"Schema\", \"Schema:gene_group_data_FB_group_name\")\n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "    return fb_id, get_mappings(q1, \"v1\")\n",
    "    \n",
    "print(f\"Mef2: {get_gene_group_names('Mef2')}\")\n",
    "print(f\"Clk: {get_gene_group_names('Clk')}\")\n",
    "print(f\"Myc: {get_gene_group_names('Myc')}\")\n",
    "print(f\"Abd-B: {get_gene_group_names('Abd-B')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "472dc645-ad10-4460-a27b-ea091be8b80a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Su[var]205: ('FBgn0003607', ['FBgg0000213'])\n",
      "Su[var]205: ('FBgn0003607', ['HP1'])\n",
      "Su[var]205: ('FBgn0003607', ['HETEROCHROMATIN PROTEIN 1 FAMILY'])\n",
      "Top3beta: ('FBgn0026015', ['FBgg0001554'])\n",
      "Top3beta: ('FBgn0026015', ['TOPI'])\n",
      "Top3beta: ('FBgn0026015', ['DNA TOPOISOMERASES TYPE I'])\n",
      "Mef2: ('FBgn0011656', ['FBgg0000753'])\n",
      "Mef2: ('FBgn0011656', ['MADSTF'])\n",
      "Mef2: ('FBgn0011656', ['MADS-BOX TRANSCRIPTION FACTORS'])\n",
      "Clk: ('FBgn0023076', ['FBgg0000727'])\n",
      "Clk: ('FBgn0023076', ['BHLH'])\n",
      "Clk: ('FBgn0023076', ['BASIC HELIX-LOOP-HELIX TRANSCRIPTION FACTORS'])\n",
      "Dref: ('FBgn0015664', ['FBgg0000732'])\n",
      "Dref: ('FBgn0015664', ['ZF-C2H2'])\n",
      "Dref: ('FBgn0015664', ['C2H2 ZINC FINGER TRANSCRIPTION FACTORS'])\n",
      "TfIIB: ('FBgn0004915', ['FBgg0000346', 'FBgg0001204'])\n",
      "TfIIB: ('FBgn0004915', ['TFIIB', 'ACTT-U'])\n",
      "TfIIB: ('FBgn0004915', ['TRANSCRIPTION FACTOR II B', 'UNCLASSIFIED ACETYLTRANSFERASES'])\n",
      "Myc: ('FBgn0262656', ['FBgg0000727'])\n",
      "Myc: ('FBgn0262656', ['BHLH'])\n",
      "Myc: ('FBgn0262656', ['BASIC HELIX-LOOP-HELIX TRANSCRIPTION FACTORS'])\n",
      "AGO2: ('FBgn0087035', ['FBgg0001553', 'FBgg0001552', 'FBgg0001778'])\n",
      "AGO2: ('FBgn0087035', ['aRLC', 'siRLC', 'AGO'])\n",
      "AGO2: ('FBgn0087035', ['ALTERNATIVE siRNA RISC-LOADING COMPLEX', 'siRNA RISC-LOADING COMPLEX', 'ARGONAUTE ENDORIBONUCLEASES'])\n",
      "Nipped-B: ('FBgn0026401', [])\n",
      "Nipped-B: ('FBgn0026401', [])\n",
      "Nipped-B: ('FBgn0026401', [])\n",
      "Cp190: ('FBgn0000283', ['FBgg0000732'])\n",
      "Cp190: ('FBgn0000283', ['ZF-C2H2'])\n",
      "Cp190: ('FBgn0000283', ['C2H2 ZINC FINGER TRANSCRIPTION FACTORS'])\n",
      "TfIIA-L: ('FBgn0011289', ['FBgg0001824', 'FBgg0001823'])\n",
      "TfIIA-L: ('FBgn0011289', ['TFIIA-S2', 'TFIIA-S'])\n",
      "TfIIA-L: ('FBgn0011289', ['GENERAL TRANSCRIPTION FACTOR II A COMPLEX - GERMLINE SPECIFIC VARIANT', 'GENERAL TRANSCRIPTION FACTOR II A COMPLEX - TFIIA-S VARIANT'])\n",
      "Trl: ('FBgn0013263', ['FBgg0000732', 'FBgg0000304'])\n",
      "Trl: ('FBgn0013263', ['ZF-C2H2', 'PCG-DNA'])\n",
      "Trl: ('FBgn0013263', ['C2H2 ZINC FINGER TRANSCRIPTION FACTORS', 'POLYCOMB GROUP RECRUITERS/DNA-BINDING PROTEINS'])\n",
      "ash1: ('FBgn0005386', ['FBgg0000274'])\n",
      "ash1: ('FBgn0005386', ['SET'])\n",
      "ash1: ('FBgn0005386', ['SET DOMAIN LYSINE METHYLTRANSFERASES'])\n",
      "Raf: ('FBgn0003079', ['FBgg0001090'])\n",
      "Raf: ('FBgn0003079', ['TKL-NR'])\n",
      "Raf: ('FBgn0003079', ['NON-RECEPTOR TKL KINASES'])\n",
      "Abd-B: ('FBgn0000015', ['FBgg0000366', 'FBgg0000747'])\n",
      "Abd-B: ('FBgn0000015', ['BX-C', 'HOXL'])\n",
      "Abd-B: ('FBgn0000015', ['BITHORAX COMPLEX', 'HOX-LIKE HOMEOBOX TRANSCRIPTION FACTORS'])\n",
      "Orc2: ('FBgn0015270', ['FBgg0000510'])\n",
      "Orc2: ('FBgn0015270', ['ORC'])\n",
      "Orc2: ('FBgn0015270', ['ORIGIN RECOGNITION COMPLEX'])\n",
      "Rbf: ('FBgn0015799', [])\n",
      "Rbf: ('FBgn0015799', [])\n",
      "Rbf: ('FBgn0015799', [])\n",
      "mof: ('FBgn0014340', ['FBgg0000283', 'FBgg0001163'])\n",
      "mof: ('FBgn0014340', ['MSL', 'HAT'])\n",
      "mof: ('FBgn0014340', ['MALE SPECIFIC LETHAL COMPLEX', 'HISTONE ACETYLTRANSFERASES'])\n",
      "msl-1: ('FBgn0005617', ['FBgg0000283'])\n",
      "msl-1: ('FBgn0005617', ['MSL'])\n",
      "msl-1: ('FBgn0005617', ['MALE SPECIFIC LETHAL COMPLEX'])\n",
      "Hmr: ('FBgn0001206', ['FBgg0000757'])\n",
      "Hmr: ('FBgn0001206', ['MADF-BESS-TR'])\n",
      "Hmr: ('FBgn0001206', ['MADF-BESS DOMAIN TRANSCRIPTION REGULATORS'])\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "This is an expanded version of the above tailored to performing queries using only the  \"essential pairing\" of columns.\n",
    "\n",
    "In fact, the three functions use the same code structure. So, the three could be merged into only one adding a \"schema parameter\"\n",
    "(and changing identifiers, of course\n",
    "\"\"\"\n",
    "# get FB id and group name(s) of gene designated by gene_symbol\n",
    "def get_gene_group_ids(gene_symbol):\n",
    "    fb_id = get_feature_fb_id(gene_symbol)    \n",
    "    n1 = Node(\"Verbatim\", fb_id)\n",
    "    v1 = Variable(\"v1\")\n",
    "    # groups\n",
    "    s = Node(\"Schema\", \"Schema:gene_group_data_FB_group_id\")\n",
    "    \n",
    "    q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "    return fb_id, get_mappings(q1, \"v1\")\n",
    "\n",
    "\n",
    "def get_gene_group_symbols(gene_symbol):\n",
    "    gene_fb_id, gene_groups_ids = get_gene_group_ids(gene_symbol)\n",
    "\n",
    "    gg_symbols = []\n",
    "    for gg_id in gene_groups_ids:\n",
    "        n1 = Node(\"Verbatim\", gg_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        # groups\n",
    "        s = Node(\"Schema\", \"Schema:gene_group_data_FB_group_symbol\")\n",
    "        \n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        for symb in get_mappings(q1, \"v1\"):\n",
    "            gg_symbols.append(symb)\n",
    "    return gene_fb_id, gg_symbols\n",
    "\n",
    "\n",
    "# get FB id and group name(s) of gene designated by gene_symbol\n",
    "def get_gene_group_names(gene_symbol):\n",
    "    gene_fb_id, gene_groups_ids = get_gene_group_ids(gene_symbol)\n",
    "    \n",
    "    gg_names = []\n",
    "    for gg_id in gene_groups_ids:\n",
    "        n1 = Node(\"Verbatim\", gg_id)\n",
    "        v1 = Variable(\"v1\")\n",
    "        # groups\n",
    "        s = Node(\"Schema\", \"Schema:gene_group_data_FB_group_name\")\n",
    "        \n",
    "        q1 = Link(\"Execution\", ordered=True, targets=[s, n1, v1])\n",
    "        for gg_name in get_mappings(q1, \"v1\"):\n",
    "            gg_names.append(gg_name)\n",
    "    return gene_fb_id, gg_names\n",
    "\n",
    "\n",
    "symbols_list = [\"Mef2\", \"Clk\", \"Myc\", \"Abd-B\"]\n",
    "symbols_list = [\"Su(var)205\", \"Top3beta\", \"Mef2\", \"Clk\", \"Dref\", \"TfIIB\", \"Myc\", \"AGO2\", \"Nipped-B\", \"Cp190\", \"TfIIA-L\",\n",
    "                \"Trl\", \"ash1\", \"Raf\", \"Abd-B\", \"Orc2\", \"Rbf\", \"mof\", \"msl-1\", \"Hmr\"]\n",
    "for gene_symbol in symbols_list:\n",
    "    gene_symbol = gene_symbol.replace(\"(\", \"[\").replace(\")\", \"]\")\n",
    "    print(f\"{gene_symbol}: {get_gene_group_ids(gene_symbol)}\")\n",
    "    print(f\"{gene_symbol}: {get_gene_group_symbols(gene_symbol)}\")\n",
    "    print(f\"{gene_symbol}: {get_gene_group_names(gene_symbol)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9e56973-9598-420b-baa9-4df04f4c41e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
